From 65df61bf369b84c9fbd2145242d09e8a854773e0 Mon Sep 17 00:00:00 2001
From: Bo-Cun Chen <bc-bocun.chen@mediatek.com>
Date: Tue, 28 May 2024 14:44:58 +0800
Subject: [PATCH] flow offload binding when there is no ARP

---
 net/netfilter/xt_FLOWOFFLOAD.c       | 49 ++++++++++++++++++++--------
 1 files changed, 34 insertions(+), 15 deletions(-)

diff --git a/net/netfilter/xt_FLOWOFFLOAD.c b/net/netfilter/xt_FLOWOFFLOAD.c
index 99fc6a1..fa10d02 100644
--- a/net/netfilter/xt_FLOWOFFLOAD.c
+++ b/net/netfilter/xt_FLOWOFFLOAD.c
@@ -395,34 +395,53 @@ static int nf_dev_fill_forward_path(const struct nf_flow_route *route,
 	if (!nf_is_valid_ether_device(dev))
 		goto out;
 
-	n = dst_neigh_lookup(dst_cache, daddr);
-	if (!n)
-		return -1;
+	if (ct->status & IPS_NAT_MASK || ct->inet6_mode == CT_INET_MODE_IPV6) {
+		n = dst_neigh_lookup(dst_cache, daddr);
+		if (!n)
+			return -1;
 
-	read_lock_bh(&n->lock);
-	nud_state = n->nud_state;
-	ether_addr_copy(ha, n->ha);
-	read_unlock_bh(&n->lock);
-	neigh_release(n);
+		read_lock_bh(&n->lock);
+		nud_state = n->nud_state;
+		ether_addr_copy(ha, n->ha);
+		read_unlock_bh(&n->lock);
+		neigh_release(n);
 
-	if (!(nud_state & NUD_VALID))
-		return -1;
+		if (!(nud_state & NUD_VALID))
+			return -1;
+	}
 
 out:
 	return dev_fill_forward_path(dev, ha, stack);
 }
 
-static int nf_dev_forward_path(struct nf_flow_route *route,
+static int nf_dev_forward_path(struct sk_buff *skb,
+				struct nf_flow_route *route,
 				const struct nf_conn *ct,
 				enum ip_conntrack_dir dir,
 				struct net_device **devs)
 {
 	const struct dst_entry *dst = route->tuple[dir].dst;
+	struct ethhdr *eth;
+	enum ip_conntrack_dir skb_dir;
 	struct net_device_path_stack stack;
 	struct nf_forward_info info = {};
 	unsigned char ha[ETH_ALEN];
 	int i;
 
+	if (!(ct->status & IPS_NAT_MASK) && skb_mac_header_was_set(skb) &&
+	    ct->inet6_mode != CT_INET_MODE_IPV6) {
+		eth = eth_hdr(skb);
+		skb_dir = CTINFO2DIR(skb_get_nfct(skb) & NFCT_INFOMASK);
+
+		if (skb_dir != dir) {
+			memcpy(ha, eth->h_source, ETH_ALEN);
+			memcpy(info.h_source, eth->h_dest, ETH_ALEN);
+		} else {
+			memcpy(ha, eth->h_dest, ETH_ALEN);
+			memcpy(info.h_source, eth->h_source, ETH_ALEN);
+		}
+	}
+
 	if (nf_dev_fill_forward_path(route, dst, ct, dir, ha, &stack) >= 0)
 		nf_dev_path_info(&stack, &info, ha);
 
@@ -512,9 +531,9 @@ xt_flowoffload_route_nat(struct sk_buff *skb, const struct nf_conn *ct,
 
 	if (route->tuple[dir].xmit_type	== FLOW_OFFLOAD_XMIT_NEIGH &&
 	    route->tuple[!dir].xmit_type == FLOW_OFFLOAD_XMIT_NEIGH) {
-		if (nf_dev_forward_path(route, ct, dir, devs))
+		if (nf_dev_forward_path(skb, route, ct, dir, devs))
 			return -1;
-		if (nf_dev_forward_path(route, ct, !dir, devs))
+		if (nf_dev_forward_path(skb, route, ct, !dir, devs))
 			return -1;
 	}
 
@@ -543,8 +562,8 @@ xt_flowoffload_route_bridge(struct sk_buff *skb, const struct nf_conn *ct,
 
 	if (route->tuple[dir].xmit_type	== FLOW_OFFLOAD_XMIT_NEIGH &&
 	    route->tuple[!dir].xmit_type == FLOW_OFFLOAD_XMIT_NEIGH) {
-		if (nf_dev_forward_path(route, ct, dir, devs) ||
-		    nf_dev_forward_path(route, ct, !dir, devs)) {
+		if (nf_dev_forward_path(skb, route, ct, dir, devs) ||
+		    nf_dev_forward_path(skb, route, ct, !dir, devs)) {
 			ret = -1;
 			goto err_route_dir2;
 		}
-- 
2.18.0

