--- a/target/linux/mediatek/image/filogic.mk	2025-04-21 16:53:40.170317466 +0800
+++ b/target/linux/mediatek/image/filogic.mk	2025-04-21 16:54:04.067371485 +0800
@@ -634,7 +634,7 @@
   DEVICE_DTS := mt7988a-bananapi-bpi-r4-poe
   DEVICE_DTS_CONFIG := config-mt7988a-bananapi-bpi-r4-poe
   $(call Device/bananapi_bpi-r4-common)
-  DEVICE_PACKAGES += mt7988-2p5g-phy-firmware
+  DEVICE_PACKAGES += mt798x-2p5g-phy-firmware-internal
 endef
 TARGET_DEVICES += bananapi_bpi-r4-poe
 
