--- a/target/linux/mediatek/image/filogic.mk
+++ b/target/linux/mediatek/image/filogic.mk
@@ -443,6 +443,7 @@ define Device/bananapi_bpi-r4-lite
   DEVICE_DTS := mt7987a-bananapi-bpi-r4-lite
   DEVICE_DTS_OVERLAY:= \
 	mt7987-spim-nand \
+	mt7987-spidev \
 	mt7987-spim-nor \
 	mt7987-emmc \
 	mt7987-sd \
@@ -477,6 +478,7 @@ define Device/mediatek_mt7987a-rfb
   DEVICE_DTS := mt7987a-rfb
   DEVICE_DTS_OVERLAY:= \
 	mt7987-spim-nand \
+	mt7987-spidev \
 	mt7987-spim-nor \
 	mt7987-emmc \
 	mt7987-sd \
