--- a/target/linux/mediatek/image/filogic.mk
+++ b/target/linux/mediatek/image/filogic.mk
@@ -447,6 +447,7 @@ define Device/mediatek_mt7987a-rfb
   DEVICE_DTS_OVERLAY:= \
 	mt7987-spim-nand \
 	mt7987-spim-nand-nmbm \
+	mt7987-spidev \
 	mt7987-spim-nor \
 	mt7987-emmc \
 	mt7987-sd \
@@ -485,6 +486,7 @@ define Device/bananapi_bpi-r4-lite
   DEVICE_DTS_OVERLAY:= \
 	mt7987-spim-nand \
 	mt7987-spim-nand-nmbm \
+	mt7987-spidev \
 	mt7987-spim-nor \
 	mt7987-emmc \
 	mt7987-sd \
