--- a/target/linux/mediatek/image/filogic-extra.mk
+++ b/target/linux/mediatek/image/filogic-extra.mk
@@ -67,7 +67,8 @@ define Device/mediatek_mt7987a-rfb
 	mt7987-netsys-eth2-an8801sb \
 	mt7987-netsys-eth2-e2p5g \
 	mt7987-netsys-eth2-sfp \
-	mt7987-netsys-eth2-usb
+	mt7987-netsys-eth2-usb \
+	mt7987-spidev
   DEVICE_DTS_DIR := $(DTS_DIR)/
   DEVICE_DTC_FLAGS := --pad 4096
   DEVICE_DTS_LOADADDR := 0x4ff00000
@@ -95,7 +96,8 @@ define Device/bananapi_bpi-r4-lite
 	mt7987-emmc \
 	mt7987-sd \
 	mt7987a-bananapi-bpi-r4-lite-1pcie-2L \
-	mt7987a-bananapi-bpi-r4-lite-2pcie-1L
+	mt7987a-bananapi-bpi-r4-lite-2pcie-1L \
+	mt7987-spidev
   DEVICE_DTS_DIR := $(DTS_DIR)/
   DEVICE_DTC_FLAGS := --pad 4096
   DEVICE_DTS_LOADADDR := 0x4ff00000
