--- a/target/linux/mediatek/image/Makefile
+++ b/target/linux/mediatek/image/Makefile
@@ -51,7 +51,18 @@ define Build/append-opteenode
 	$(STAGING_DIR_HOSTPKG)/bin/fdtput $(KDIR)/image-$(firstword $(DEVICE_DTS)).dtb /firmware/optee -ts status "okay"
 endef
 
+# build squashfs-hashed
+define Build/squashfs-hashed
+	$(TOPDIR)/scripts/make-squashfs-hashed.sh \
+		$(ROOTFS/$(FILESYSTEMS)/$(DEVICE_NAME)) \
+		$(STAGING_DIR_HOSTPKG) \
+		$(ROOTFS/$(FILESYSTEMS)/$(DEVICE_NAME))-hashed-summary
+	fdt-patch-dm-verify $(ROOTFS/$(FILESYSTEMS)/$(DEVICE_NAME))-hashed-summary \
+		$(KDIR)/image-$(firstword $(DEVICE_DTS)).dtb $(KDIR)/image-$(firstword $(DEVICE_DTS)).dtb
+endef
+
 define Build/secure-boot
+	$(if $(CONFIG_MTK_SECURE_BOOT),$(call Build/squashfs-hashed))
 	$(if $(CONFIG_PACKAGE_optee-mediatek),$(call Build/append-opteenode))
 endef
 
