--- a/target/linux/mediatek/image/Makefile
+++ b/target/linux/mediatek/image/Makefile
@@ -33,6 +33,17 @@ define Device/Default
 	pad-rootfs | append-metadata
 endef
 
+define Image/get_fw_ar_ver
+	$(shell sed -n 's/^FW_AR_VER\t:=\t*//p' $(TOPDIR)/$(call qstrip,$(CONFIG_AUTO_AR_CONF)))
+endef
+
+define Image/gen-fw-ar-ver
+	rm -rf $(TOPDIR)/$(call qstrip,$(CONFIG_AUTO_AR_CONF))
+	$(STAGING_DIR_HOST)/bin/ar-tool fw_ar_table create_ar_conf \
+		$(TOPDIR)/$(call qstrip,$(CONFIG_ANTI_ROLLBACK_TABLE)) \
+		$(TOPDIR)/$(call qstrip,$(CONFIG_AUTO_AR_CONF))
+endef
+
 include $(SUBTARGET).mk
 -include $(SUBTARGET)-extra.mk
 
--- a/target/linux/mediatek/image/filogic.mk
+++ b/target/linux/mediatek/image/filogic.mk
@@ -5,6 +5,7 @@ define Image/Prepare
 	rm -f $(KDIR)/ubi_mark
 	echo -ne '\xde\xad\xc0\xde' > $(KDIR)/ubi_mark
 	$(if $(CONFIG_MTK_FW_ENC),$(call Image/fw-enc-key-derive))
+	$(if $(CONFIG_MTK_ANTI_ROLLBACK),$(call Image/gen-fw-ar-ver))
 endef
 
 define Build/mt7981-bl2
