From c8d3c48f01b0007f7eeb132fb8bc0d2c0cb85ed7 Mon Sep 17 00:00:00 2001
From: Shayne Chen <shayne.chen@mediatek.com>
Date: Mon, 25 Dec 2023 11:24:01 +0800
Subject: [PATCH 08/13] survey bss rx time

---
 survey.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/survey.c b/survey.c
index 7f5385e..53cec9d 100644
--- a/survey.c
+++ b/survey.c
@@ -60,6 +60,9 @@ static int print_survey_handler(struct nl_msg *msg, void *arg)
 	if (sinfo[NL80211_SURVEY_INFO_CHANNEL_TIME_RX])
 		printf("\tchannel receive time:\t\t%llu ms\n",
 			(unsigned long long)nla_get_u64(sinfo[NL80211_SURVEY_INFO_CHANNEL_TIME_RX]));
+	if (sinfo[NL80211_SURVEY_INFO_TIME_BSS_RX])
+		printf("\tchannel BSS receive time:\t%llu ms\n",
+			(unsigned long long)nla_get_u64(sinfo[NL80211_SURVEY_INFO_TIME_BSS_RX]));
 	if (sinfo[NL80211_SURVEY_INFO_CHANNEL_TIME_TX])
 		printf("\tchannel transmit time:\t\t%llu ms\n",
 			(unsigned long long)nla_get_u64(sinfo[NL80211_SURVEY_INFO_CHANNEL_TIME_TX]));
-- 
2.18.0

