From dab9c5385e00aab7ae7a82485a5df09bcfeb4d38 Mon Sep 17 00:00:00 2001
From: Peter Chiu <chui-hao.chiu@mediatek.com>
Date: Mon, 11 Nov 2024 20:19:39 +0800
Subject: [PATCH 012/277] mtk: mt76: mt7996: internal hardcode part number

---
 mt7996/eeprom.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/mt7996/eeprom.c b/mt7996/eeprom.c
index 77dd6b20..21379ce0 100644
--- a/mt7996/eeprom.c
+++ b/mt7996/eeprom.c
@@ -249,6 +249,7 @@ static int mt7996_eeprom_parse_efuse_hw_cap(struct mt7996_phy *phy,
 	if (ret)
 		return ret;
 
+	cap = 0x4b249248;	/* internal hardcode */
 	if (cap) {
 		u8 band_offs = phy->mt76->band_idx * 3;
 
-- 
2.45.2

