From cf36a04f5af428f1145b0c04fe92090e29be74a3 Mon Sep 17 00:00:00 2001
From: MeiChia Chiu <meichia.chiu@mediatek.com>
Date: Tue, 1 Aug 2023 16:02:28 +0800
Subject: [PATCH 029/277] mtk: mt76: mt7996: Establish BA in VO queue

---
 mt7996/mac.c | 2 --
 1 file changed, 2 deletions(-)

diff --git a/mt7996/mac.c b/mt7996/mac.c
index e9f3942f..428ccb1a 100644
--- a/mt7996/mac.c
+++ b/mt7996/mac.c
@@ -1053,8 +1053,6 @@ mt7996_tx_check_aggr(struct ieee80211_sta *sta, struct sk_buff *skb)
 		return;
 
 	tid = skb->priority & IEEE80211_QOS_CTL_TID_MASK;
-	if (tid >= 6) /* skip VO queue */
-		return;
 
 	if (is_8023) {
 		fc = IEEE80211_FTYPE_DATA |
-- 
2.45.2

