From d56b07bfef1328ee9ce756a3d69360e6a8dc653a Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Thu, 2 Jan 2025 09:27:35 +0800
Subject: [PATCH 52/95] mtk: mt76: mt7996: set TTLM negotiation support flag

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 mt7996/init.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/mt7996/init.c b/mt7996/init.c
index 54c352fa..ad7477c2 100644
--- a/mt7996/init.c
+++ b/mt7996/init.c
@@ -91,13 +91,17 @@ static const u8 mt7996_if_types_ext_capa_sta[] = {
 	[7] = WLAN_EXT_CAPA8_OPMODE_NOTIF,
 };
 
+#define MT7996_NEG_TTLM_SUPPORT FIELD_PREP_CONST( \
+		IEEE80211_MLD_CAP_OP_TID_TO_LINK_MAP_NEG_SUPP, \
+		IEEE80211_MLD_CAP_OP_TID_TO_LINK_MAP_NEG_SUPP_DIFF)
+
 static const struct wiphy_iftype_ext_capab mt7996_iftypes_ext_capa[] = {
 	{
 		.iftype = NL80211_IFTYPE_STATION,
 		.extended_capabilities = mt7996_if_types_ext_capa_sta,
 		.extended_capabilities_mask = mt7996_if_types_ext_capa_sta,
 		.extended_capabilities_len = sizeof(mt7996_if_types_ext_capa_sta),
-		.mld_capa_and_ops = 2,
+		.mld_capa_and_ops = 2 | MT7996_NEG_TTLM_SUPPORT,
 	},
 	{
 		.iftype = NL80211_IFTYPE_AP,
@@ -105,7 +109,7 @@ static const struct wiphy_iftype_ext_capab mt7996_iftypes_ext_capa[] = {
 		.extended_capabilities_mask = mt7996_if_types_ext_capa_ap,
 		.extended_capabilities_len = sizeof(mt7996_if_types_ext_capa_ap),
 		.eml_capabilities = IEEE80211_EML_CAP_EMLSR_SUPP,
-		.mld_capa_and_ops = 2,
+		.mld_capa_and_ops = 2 | MT7996_NEG_TTLM_SUPPORT,
 		/* the max number of simultaneous links is defined as the
 		 * maximum number of affiliated APs minus 1.
 		 * mt7996 could have 3 links in an MLD AP, so currently
-- 
2.45.2

