From 5ee23ceb09973820320cfb40bf61b32f13333229 Mon Sep 17 00:00:00 2001
From: Peter Chiu <chui-hao.chiu@mediatek.com>
Date: Mon, 19 Jan 2026 09:22:05 +0800
Subject: [PATCH 2/4] mtk: mt76: mt7996: fix incorrect queue id for debug code

Signed-off-by: Peter Chiu <chui-hao.chiu@mediatek.com>
---
 mt7996/mtk_debugfs.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/mt7996/mtk_debugfs.c b/mt7996/mtk_debugfs.c
index d05283aae..09c29d6f0 100644
--- a/mt7996/mtk_debugfs.c
+++ b/mt7996/mtk_debugfs.c
@@ -3917,10 +3917,10 @@ static struct bmac_queue_info ple_queue_empty_info[] = {
 	{"BMC Q1",  ENUM_UMAC_LMAC_PORT_2,    0x11, 1},
 	{"BCN Q1",  ENUM_UMAC_LMAC_PORT_2,    0x12, 1},
 	{"PSMP Q1", ENUM_UMAC_LMAC_PORT_2,    0x13, 1},
-	{"ALTX Q2", ENUM_UMAC_LMAC_PORT_2,    0x10, 2},
-	{"BMC Q2",  ENUM_UMAC_LMAC_PORT_2,    0x11, 2},
-	{"BCN Q2",  ENUM_UMAC_LMAC_PORT_2,    0x12, 2},
-	{"PSMP Q2", ENUM_UMAC_LMAC_PORT_2,    0x13, 2},
+	{"ALTX Q2", ENUM_UMAC_LMAC_PORT_2,    0x1c, 2},
+	{"BMC Q2",  ENUM_UMAC_LMAC_PORT_2,    0x1d, 2},
+	{"BCN Q2",  ENUM_UMAC_LMAC_PORT_2,    0x1e, 2},
+	{"PSMP Q2", ENUM_UMAC_LMAC_PORT_2,    0x1f, 2},
 	{"NAF Q",   ENUM_UMAC_LMAC_PORT_2,    0x18, 0},
 	{"NBCN Q",  ENUM_UMAC_LMAC_PORT_2,    0x19, 0},
 	{NULL, 0, 0, 0}, {NULL, 0, 0, 0}, /* 18, 19 not defined */
-- 
2.45.2

