From 5bea562ec066acd49e460e8d16f3f55795298895 Mon Sep 17 00:00:00 2001
From: Shayne Chen <shayne.chen@mediatek.com>
Date: Thu, 24 Oct 2024 14:29:09 +0800
Subject: [PATCH 237/277] fixup! mtk: mt76: mt7996: hardcode has_eht on some
 chips

---
 mt7996/eeprom.c | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/mt7996/eeprom.c b/mt7996/eeprom.c
index 80c4b0ba..6d8df3a0 100644
--- a/mt7996/eeprom.c
+++ b/mt7996/eeprom.c
@@ -391,8 +391,7 @@ static int mt7996_eeprom_parse_efuse_hw_cap(struct mt7996_phy *phy,
 		return ret;
 
 	/* for internal development */
-	if (is_mt7992(&dev->mt76))
-		dev->has_eht = true;
+	dev->has_eht = true;
 	if (cap) {
 		u8 band_offs = phy->mt76->band_idx * 3;
 
-- 
2.45.2

