From 4d685b9f5c25234b701364676b6fdab3a21dd1b1 Mon Sep 17 00:00:00 2001
From: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
Date: Wed, 13 Nov 2024 14:05:54 +0800
Subject: [PATCH 253/277] fixup! mtk: mt76: mt7996: add listmode support

Fix warning

Signed-off-by: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
---
 testmode.c | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

diff --git a/testmode.c b/testmode.c
index f1f49a2c..e1926353 100644
--- a/testmode.c
+++ b/testmode.c
@@ -506,9 +506,16 @@ mt76_testmode_set_list_mode(struct mt76_phy *phy, struct nlattr **tb)
 		if (phy->lists[0].seg_type != seg_type)
 			return -EINVAL;
 		if (seg_idx == phy->seg_num) {
+			unsigned long size;
+			void *tmp;
+
 			phy->seg_num++;
-			krealloc((void *)phy->lists, sizeof(*phy->lists) * phy->seg_num,
-				 GFP_KERNEL);
+			size = sizeof(*phy->lists) * phy->seg_num;
+			tmp = krealloc((void *)phy->lists, size, GFP_KERNEL);
+			if (!tmp)
+				return -ENOMEM;
+
+			phy->lists = (struct mt76_list_mode_data *)tmp;
 			mt76_testmode_list_init_defaults(phy, seg_idx, seg_type);
 		}
 	}
-- 
2.45.2

