From 2fcfae109e25aa96e4791545cb7cd49eb79ad898 Mon Sep 17 00:00:00 2001
From: Shayne Chen <shayne.chen@mediatek.com>
Date: Thu, 7 Nov 2024 14:39:27 +0800
Subject: [PATCH 268/277] fixup! mtk: mt76: mt7996: switch to single
 multi-radio wiphy

---
 mt7996/init.c | 18 +-----------------
 1 file changed, 1 insertion(+), 17 deletions(-)

diff --git a/mt7996/init.c b/mt7996/init.c
index 0aed2bbb..b9cd15cb 100644
--- a/mt7996/init.c
+++ b/mt7996/init.c
@@ -523,25 +523,9 @@ mt7996_init_wiphy(struct ieee80211_hw *hw, struct mtk_wed_device *wed)
 
 	wiphy->iface_combinations = &if_comb_global;
 	wiphy->n_iface_combinations = 1;
+	wiphy->mbssid_max_interfaces = is_mt7996(&dev->mt76) ? 48 : 32;
 	wiphy->radio = dev->radios;
 
-#if 0
-	switch (mt76_chip(mdev)) {
-	case MT7996_DEVICE_ID:
-		wiphy->iface_combinations = if_comb;
-		wiphy->n_iface_combinations = ARRAY_SIZE(if_comb);
-		wiphy->mbssid_max_interfaces = 16 * 3;
-		break;
-	case MT7992_DEVICE_ID:
-	case MT7990_DEVICE_ID:
-	default:
-		wiphy->iface_combinations = if_comb_7992;
-		wiphy->n_iface_combinations = ARRAY_SIZE(if_comb_7992);
-		wiphy->mbssid_max_interfaces = 16 * 2;
-		break;
-	}
-#endif
-
 	wiphy->reg_notifier = mt7996_regd_notifier;
 	wiphy->flags |= WIPHY_FLAG_HAS_CHANNEL_SWITCH;
 
-- 
2.45.2

