From b0a7f123f89c761d277e35d069a61e6eb67cf88a Mon Sep 17 00:00:00 2001
From: Shayne Chen <shayne.chen@mediatek.com>
Date: Thu, 7 Nov 2024 17:36:40 +0800
Subject: [PATCH 270/277] fixup! mtk: mt76: mt7996: switch to single
 multi-radio wiphy

---
 mt7996/init.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/mt7996/init.c b/mt7996/init.c
index 3f8905c2..d290f070 100644
--- a/mt7996/init.c
+++ b/mt7996/init.c
@@ -476,7 +476,7 @@ mt7996_init_wiphy_band(struct ieee80211_hw *hw, struct mt7996_phy *phy)
 		freq->start_freq = 5000000;
 		freq->end_freq = 5900000;
 	} else if (phy->mt76->cap.has_6ghz) {
-		freq->start_freq = 5900000;
+		freq->start_freq = 5925000;
 		freq->end_freq = 7200000;
 	} else {
 		return;
-- 
2.45.2

