From 1a5230a4cef3b76bb25a0750987a3a1229729f11 Mon Sep 17 00:00:00 2001
From: Shayne Chen <shayne.chen@mediatek.com>
Date: Wed, 13 Nov 2024 16:36:34 +0800
Subject: [PATCH 276/277] fixup! fixup! mtk: mt76: mt7996: switch to single
 multi-radio wiphy

---
 mt7996/init.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/mt7996/init.c b/mt7996/init.c
index a487b339..67790cee 100644
--- a/mt7996/init.c
+++ b/mt7996/init.c
@@ -1871,11 +1871,11 @@ error:
 void mt7996_unregister_device(struct mt7996_dev *dev)
 {
 	cancel_work_sync(&dev->wed_rro.work);
+	mt7996_unregister_phy(mt7996_phy3(dev));
+	mt7996_unregister_phy(mt7996_phy2(dev));
 	mt7996_unregister_thermal(&dev->phy);
 	mt7996_coredump_unregister(dev);
 	mt76_unregister_device(&dev->mt76);
-	mt7996_unregister_phy(mt7996_phy3(dev));
-	mt7996_unregister_phy(mt7996_phy2(dev));
 	mt7996_wed_rro_free(dev);
 	mt7996_mcu_exit(dev);
 	mt7996_tx_token_put(dev);
-- 
2.45.2

