--- a/target/linux/mediatek/files-6.6/arch/arm64/boot/dts/mediatek/mt7988a-rfb-spim-nand-factory.dtso
+++ b/target/linux/mediatek/files-6.6/arch/arm64/boot/dts/mediatek/mt7988a-rfb-spim-nand-factory.dtso
@@ -79,4 +79,11 @@
 			nvmem-cells = <&gmac2_mac>;
 		};
 	};
+
+	fragment@5 {
+		target-path = "/";
+		__overlay__ {
+			mediatek,no-split-fitrw;
+		};
+	};
 };
--- a/target/linux/mediatek/files-6.6/arch/arm64/boot/dts/mediatek/mt7988a-rfb-spim-nand.dtso
+++ b/target/linux/mediatek/files-6.6/arch/arm64/boot/dts/mediatek/mt7988a-rfb-spim-nand.dtso
@@ -72,4 +72,11 @@
 			rootdisk-spim-nand = <&ubi_root>;
 		};
 	};
+
+	fragment@2 {
+		target-path = "/";
+		__overlay__ {
+			mediatek,no-split-fitrw;
+		};
+	};
 };
