--- a/package/kernel/linux/modules/block.mk
+++ b/package/kernel/linux/modules/block.mk
@@ -232,7 +232,12 @@ define KernelPackage/dm
 	CONFIG_MD=y \
 	CONFIG_BLK_DEV_DM=y \
 	CONFIG_DM_CRYPT \
-	CONFIG_DM_MIRROR
+	CONFIG_DM_MIRROR \
+	CONFIG_DM_DEBUG_BLOCK_MANAGER_LOCKING=n \
+	CONFIG_DM_INIT=y \
+	CONFIG_DM_VERITY=y \
+	CONFIG_DM_VERITY_FEC=n \
+	CONFIG_DM_VERITY_VERIFY_ROOTHASH_SIG=n
   FILES:= \
     $(LINUX_DIR)/drivers/md/dm-crypt.ko \
     $(LINUX_DIR)/drivers/md/dm-log.ko \
