--- a/package/boot/uboot-tools/Makefile
+++ b/package/boot/uboot-tools/Makefile
@@ -29,6 +29,7 @@ define Package/dumpimage
 	CATEGORY:=Utilities
 	SUBMENU:=Boot Loaders
 	TITLE:=dumpimage lists and extracts data from U-Boot images.
+	DEPENDS:=+libopenssl
 endef
 
 define Package/dumpimage/description
@@ -43,6 +44,7 @@ define Package/fit-check-sign
 	CATEGORY:=Utilities
 	SUBMENU:=Boot Loaders
 	TITLE:=verify uImage.FIT
+	DEPENDS:=+libopenssl
 endef
 
 define Package/fit-check-sign/description
@@ -70,7 +72,6 @@ endef
 
 define Build/Configure
 	$(call Build/Compile/Default,tools-only_defconfig)
-	$(SED) 's/CONFIG_TOOLS_LIBCRYPTO=y/# CONFIG_TOOLS_LIBCRYPTO is not set/' $(PKG_BUILD_DIR)/.config
 	$(SED) 's/CONFIG_TOOLS_KWBIMAGE=y/# CONFIG_TOOLS_KWBIMAGE is not set/' $(PKG_BUILD_DIR)/.config
 endef
 
