diff --git a/package/kernel/mt76/Makefile b/package/kernel/mt76/Makefile
index 9a8b64d..1acc138 100644
--- a/package/kernel/mt76/Makefile
+++ b/package/kernel/mt76/Makefile
@@ -8,9 +8,9 @@ PKG_LICENSE_FILES:=
 
 PKG_SOURCE_URL:=https://github.com/openwrt/mt76
 PKG_SOURCE_PROTO:=git
-PKG_SOURCE_DATE:=2024-07-13
-PKG_SOURCE_VERSION:=3b47d9df427c4833605a172f2a8f0e0012b04c80
-PKG_MIRROR_HASH:=23c3aaa53fb2e088446eb18148a44d3edcd3a0eda1ee41cf5cbf56064ebbee58
+PKG_SOURCE_DATE:=2024-07-16
+PKG_SOURCE_VERSION:=6585a4353a51c0e4873b691bbcab251dffc7e2bb
+PKG_MIRROR_HASH:=xxxxx
 
 PKG_MAINTAINER:=Felix Fietkau <nbd@nbd.name>
 PKG_USE_NINJA:=0

