diff --git a/package/utils/ucode/Makefile b/package/utils/ucode/Makefile
index ef253f89..1f4f4065 100644
--- a/package/utils/ucode/Makefile
+++ b/package/utils/ucode/Makefile
@@ -12,9 +12,9 @@ PKG_RELEASE:=1
 
 PKG_SOURCE_PROTO:=git
 PKG_SOURCE_URL=https://github.com/jow-/ucode.git
-PKG_SOURCE_DATE:=2025-02-10
-PKG_SOURCE_VERSION:=a8a11aea0c093d669bb3c45f604dab3c291c8d25
-PKG_MIRROR_HASH:=e08c36818a72eb2953b56300207d1fb7a852ada33df36f63a495ad35ca21b88e
+PKG_SOURCE_DATE:=2024-07-22
+PKG_SOURCE_VERSION:=b610860dd4a0591ff586dd71a50f556a0ddafced
+PKG_MIRROR_HASH:=a5ec51dd989174422d3b19b022ff4f863d57eb559c9f08d54c0d10651f598357
 PKG_MAINTAINER:=Jo-Philipp Wich <jo@mein.io>
 PKG_LICENSE:=ISC
 
