diff --git a/package/network/services/hostapd/Makefile b/package/network/services/hostapd/Makefile
index e40de4c..bfd2371 100644
--- a/package/network/services/hostapd/Makefile
+++ b/package/network/services/hostapd/Makefile
@@ -9,9 +9,9 @@ PKG_RELEASE:=1
 
 PKG_SOURCE_URL:=https://w1.fi/hostap.git
 PKG_SOURCE_PROTO:=git
-PKG_SOURCE_DATE:=2024-03-09
-PKG_SOURCE_VERSION:=695277a5b3da08b9a8a4e7117b933deb8b4950a7
-PKG_MIRROR_HASH:=57b8e64d24707e37e0df3359cee15dd5184b824b8622568833a5b8a0cae163ae
+PKG_SOURCE_DATE:=2024-03-27
+PKG_SOURCE_VERSION:=07c9f183ea744ac04585fb6dd10220c75a5e2e74
+PKG_MIRROR_HASH:=db92f2ac59039fe02d4c47d1ca314f7f251d0d5689876e19fbf4f72a4111f39c
 
 PKG_MAINTAINER:=Felix Fietkau <nbd@nbd.name>
 PKG_LICENSE:=BSD-3-Clause
