diff --git a/package/network/services/hostapd/Makefile b/package/network/services/hostapd/Makefile
index 9692c23..bcd10f6 100644
--- a/package/network/services/hostapd/Makefile
+++ b/package/network/services/hostapd/Makefile
@@ -9,9 +9,9 @@ PKG_RELEASE:=3
 
 PKG_SOURCE_URL:=https://w1.fi/hostap.git
 PKG_SOURCE_PROTO:=git
-PKG_SOURCE_DATE:=2024-03-09
-PKG_SOURCE_VERSION:=695277a5b3da08b9a8a4e7117b933deb8b4950a7
-PKG_MIRROR_HASH:=57b8e64d24707e37e0df3359cee15dd5184b824b8622568833a5b8a0cae163ae
+PKG_SOURCE_DATE:=2024-07-31
+PKG_SOURCE_VERSION:=7e0e69cfeac300414ef0492bc76a2aa164443249
+PKG_MIRROR_HASH:=db92f2ac59039fe02d4c47d1ca314f7f251d0d5689876e19fbf4f72a4111f39c
 
 PKG_MAINTAINER:=Felix Fietkau <nbd@nbd.name>
 PKG_LICENSE:=BSD-3-Clause
