diff --git a/package/network/services/hostapd/Makefile b/package/network/services/hostapd/Makefile
index 9692c23..bcd10f6 100644
--- a/package/network/services/hostapd/Makefile
+++ b/package/network/services/hostapd/Makefile
@@ -9,9 +9,9 @@ PKG_RELEASE:=3
 
 PKG_SOURCE_URL:=https://w1.fi/hostap.git
 PKG_SOURCE_PROTO:=git
-PKG_SOURCE_DATE:=2024-09-15
-PKG_SOURCE_VERSION:=5ace39b0a4cdbe18ddbc4e18f80ee3876233c20b
-PKG_MIRROR_HASH:=43187d1ddd923970ef3821a0fe267880c2483f6f1a5acd0dd2d7906661ab1372
+PKG_SOURCE_DATE:=2024-07-31
+PKG_SOURCE_VERSION:=7e0e69cfeac300414ef0492bc76a2aa164443249
+PKG_MIRROR_HASH:=4607c3b215184734b62766fde07f2e1e8c20e078c68e5ea4cf9097ad52cd4e00
 
 PKG_MAINTAINER:=Felix Fietkau <nbd@nbd.name>
 PKG_LICENSE:=BSD-3-Clause
