diff --git a/package/network/services/hostapd/files/hostapd.uc b/package/network/services/hostapd/files/hostapd.uc
index 2bc58e4..9e77cde 100644
--- a/package/network/services/hostapd/files/hostapd.uc
+++ b/package/network/services/hostapd/files/hostapd.uc
@@ -53,7 +53,8 @@ function iface_remove(cfg)
 		return;
 
 	for (let bss in cfg.bss)
-		wdev_remove(bss.ifname);
+		if (!bss.mld_ap || bss.mld_primary == 1)
+			wdev_remove(bss.ifname);
 }
 
 function iface_gen_config(config, start_disabled)
@@ -87,6 +88,7 @@ function iface_freq_info(iface, config, params)
 {
 	let freq = params.frequency;
 	let bw320_offset = params.bw320_offset;
+	let band_idx = params.band_idx;
 	if (!freq)
 		return null;
 
@@ -117,7 +119,7 @@ function iface_freq_info(iface, config, params)
 	if (freq < 4000)
 		width = 0;
 
-	return hostapd.freq_info(freq, sec_offset, width, bw320_offset);
+	return hostapd.freq_info(freq, sec_offset, width, bw320_offset, band_idx);
 }
 
 function iface_add(phy, config, phy_status)
@@ -180,17 +182,19 @@ function __iface_pending_next(pending, state, ret, data)
 		iface_update_supplicant_macaddr(phydev, config);
 		return "create_bss";
 	case "create_bss":
-		let err = phydev.wdev_add(bss.ifname, {
-			mode: "ap",
-			radio: phydev.radio,
-		});
-		if (err) {
-			hostapd.printf(`Failed to create ${bss.ifname} on phy ${phy}: ${err}`);
-			return null;
+		if (!bss.mld_ap || bss.mld_primary == 1) {
+			let err = phydev.wdev_add(bss.ifname, {
+				mode: "ap",
+				radio: phydev.radio,
+			});
+			if (err) {
+				hostapd.printf(`Failed to create ${bss.ifname} on phy ${phy}: ${err}`);
+				return null;
+			}
 		}
 
 		pending.call("wpa_supplicant", "phy_status", {
-			phy: phydev.phy,
+			phy: bss.mld_ap ? "phy0" : phydev.phy,
 			radio: phydev.radio,
 		});
 		return "check_phy";
@@ -203,7 +207,7 @@ function __iface_pending_next(pending, state, ret, data)
 			hostapd.printf(`Failed to bring up phy ${phy} ifname=${bss.ifname} with supplicant provided frequency`);
 		}
 		pending.call("wpa_supplicant", "phy_set_state", {
-			phy: phydev.phy,
+			phy: bss.mld_ap ? "phy0" : phydev.phy,
 			radio: phydev.radio,
 			stop: true
 		});
@@ -211,11 +215,20 @@ function __iface_pending_next(pending, state, ret, data)
 	case "wpas_stopped":
 		if (!iface_add(phy, config))
 			hostapd.printf(`hostapd.add_iface failed for phy ${phy} ifname=${bss.ifname}`);
-		pending.call("wpa_supplicant", "phy_set_state", {
-			phy: phydev.phy,
-			radio: phydev.radio,
-			stop: false
-		});
+		let iface = hostapd.interfaces[phy];
+		if (!bss.mld_ap) {
+			pending.call("wpa_supplicant", "phy_set_state", {
+				phy: phydev.phy,
+				radio: phydev.radio,
+				stop: false
+			});
+		} else if (iface.is_mld_finished()) {
+			pending.call("wpa_supplicant", "phy_set_state", {
+				phy: "phy0",
+				radio: phydev.radio,
+				stop: false
+			});
+		}
 		return null;
 	case "done":
 	default:
@@ -754,6 +767,9 @@ function iface_load_config(phy, radio, filename)
 		    val[0] == "mbssid")
 			config[substr(val[0], 1)] = int(val[1]);
 
+		if (val[0] == "#single_hw")
+			config["single_hw"] = int(val[1]);
+
 		push(config.radio.data, line);
 	}
 
@@ -770,6 +786,12 @@ function iface_load_config(phy, radio, filename)
 			continue;
 		}
 
+		if (val[0] == "mld_ap" && int(val[1]) == 1)
+			bss.mld_ap = 1;
+
+		if (val[0] == "mld_primary" && int(val[1]) == 1)
+			bss.mld_primary = 1;
+
 		if (val[0] == "nas_identifier")
 			bss.nasid = val[1];
 
@@ -785,6 +807,19 @@ function iface_load_config(phy, radio, filename)
 	}
 	f.close();
 
+	let first_mld_bss = 0;
+	for (first_mld_bss = 0; first_mld_bss < length(config.bss); first_mld_bss++) {
+		if (config.bss[first_mld_bss].mld_ap == 1)
+			break;
+	}
+
+	if (config.bss[0].mld_ap != 1 && first_mld_bss != length(config.bss)) {
+		let tmp_bss = config.bss[0];
+		config.bss[0] = config.bss[first_mld_bss];
+		config.bss[first_mld_bss] = tmp_bss;
+		hostapd.printf(`mtk: ucode: switch bss[${first_mld_bss}] to first`);
+	}
+
 	return config;
 }
 
@@ -848,6 +883,7 @@ let main_obj = {
 			sec_chan_offset: 0,
 			ch_width: -1,
 			bw320_offset: 1,
+			band_idx: 0,
 			csa: true,
 			csa_count: 0,
 		},
@@ -863,6 +899,7 @@ let main_obj = {
 			hostapd.printf(`    * sec_chan_offset: ${req.args.sec_chan_offset}`);
 			hostapd.printf(`    * ch_width: ${req.args.ch_width}`);
 			hostapd.printf(`    * bw320_offset: ${req.args.bw320_offset}`);
+			hostapd.printf(`    * band_idx: ${req.args.band_idx}`);
 			hostapd.printf(`    * csa: ${req.args.csa}`);
 
 			let config = hostapd.data.config[phy];
diff --git a/package/network/services/hostapd/files/wpa_supplicant.uc b/package/network/services/hostapd/files/wpa_supplicant.uc
index a9b7f9d..aae7acc 100644
--- a/package/network/services/hostapd/files/wpa_supplicant.uc
+++ b/package/network/services/hostapd/files/wpa_supplicant.uc
@@ -310,10 +310,9 @@ function iface_event(type, name, data) {
 	ubus.call("service", "event", { type: `wpa_supplicant.${name}.${type}`, data: {} });
 }
 
-function iface_hostapd_notify(phy, ifname, iface, state)
+function iface_hostapd_notify(phy, ifname, iface, state, link_id)
 {
 	let ubus = wpas.data.ubus;
-	let status = iface.status();
 	let msg = { phy: phy };
 
 	wpas.printf(`ucode: mtk: wpa_s in state ${state} notifies hostapd`);
@@ -328,11 +327,13 @@ function iface_hostapd_notify(phy, ifname, iface, state)
 		msg.up = true;
 		break;
 	case "COMPLETED":
+		let status = iface.status(link_id);
 		msg.up = true;
 		msg.frequency = status.frequency;
 		msg.sec_chan_offset = status.sec_chan_offset;
 		msg.ch_width = status.ch_width;
 		msg.bw320_offset = status.bw320_offset;
+		msg.band_idx = status.band_idx;
 		break;
 	default:
 		return;
@@ -351,6 +352,7 @@ function iface_channel_switch(phy, ifname, iface, info)
 		frequency: info.frequency,
 		ch_width: info.ch_width,
 		bw320_offset: info.bw320_offset,
+		band_idx: info.band_idx,
 		sec_chan_offset: info.sec_chan_offset,
 	};
 	ubus.call("hostapd", "apsta_state", msg);
@@ -370,12 +372,24 @@ return {
 	},
 	state: function(ifname, iface, state) {
 		let phy = wpas.data.iface_phy[ifname];
+		let ret = iface.get_valid_links();
+		let link_id = 0, valid_links = ret.valid_links;
 		if (!phy) {
 			wpas.printf(`no PHY for ifname ${ifname}`);
 			return;
 		}
 
-		iface_hostapd_notify(phy, ifname, iface, state);
+		if (valid_links) {
+			while (valid_links) {
+				if (valid_links & 1)
+					iface_hostapd_notify(phy, ifname, iface, state, link_id);
+
+				link_id++;
+				valid_links >>= 1;
+			}
+		} else {
+			iface_hostapd_notify(phy, ifname, iface, state, -1);
+		}
 
 		if (state != "COMPLETED")
 			return;
@@ -401,7 +415,7 @@ return {
 			return;
 		}
 
-		if (ev == "CH_SWITCH_STARTED")
+		if (ev == "CH_SWITCH_STARTED" || ev == "LINK_CH_SWITCH_STARTED")
 			iface_channel_switch(phy, ifname, iface, info);
 	}
 };
