diff --git a/package/libs/libbpf/Makefile b/package/libs/libbpf/Makefile
index 89fe730a98..0f1fa1db97 100644
--- a/package/libs/libbpf/Makefile
+++ b/package/libs/libbpf/Makefile
@@ -12,7 +12,7 @@ PKG_VERSION:=1.4.6
 PKG_RELEASE:=1
 
 PKG_SOURCE_URL:=https://github.com/libbpf/libbpf
-PKG_MIRROR_HASH:=0d0d70750c8e45baba19f8cc065771e5d842a90f3ad1637003cae6887dabf11e
+PKG_MIRROR_HASH:=0c9d1b842d36b451e1d104d8b825508595b49ed6cbfa44dc46b2cba8f1eded61
 PKG_SOURCE_PROTO:=git
 PKG_SOURCE_VERSION:=v$(PKG_VERSION)
 PKG_ABI_VERSION:=$(firstword $(subst .,$(space),$(PKG_VERSION)))
diff --git a/package/libs/libnl-tiny/Makefile b/package/libs/libnl-tiny/Makefile
index c211288f8e..a0c6efe517 100644
--- a/package/libs/libnl-tiny/Makefile
+++ b/package/libs/libnl-tiny/Makefile
@@ -14,7 +14,7 @@ PKG_SOURCE_PROTO:=git
 PKG_SOURCE_URL=$(PROJECT_GIT)/project/libnl-tiny.git
 PKG_SOURCE_DATE:=2023-12-05
 PKG_SOURCE_VERSION:=965c4bf49658342ced0bd6e7cb069571b4a1ddff
-PKG_MIRROR_HASH:=4b56499e71f5f10b123e1c7ee13dbfcd970de560bff8d16ab39dc25c9ff743cb
+PKG_MIRROR_HASH:=93596c2686926a470b6bf322f6c166a420a08abe1ea715b4ae8c9df3de00eb28
 CMAKE_INSTALL:=1
 
 PKG_LICENSE:=LGPL-2.1
diff --git a/package/libs/libubox/Makefile b/package/libs/libubox/Makefile
index 22abfd4372..e77dfd9874 100644
--- a/package/libs/libubox/Makefile
+++ b/package/libs/libubox/Makefile
@@ -5,7 +5,7 @@ PKG_RELEASE=1
 
 PKG_SOURCE_PROTO:=git
 PKG_SOURCE_URL=$(PROJECT_GIT)/project/libubox.git
-PKG_MIRROR_HASH:=a4f671d10840fd8487394335636051a6df5edf6d8854af4fcb834a590efb240a
+PKG_MIRROR_HASH:=242e33eca235124c7e005d25fbc8f8bf08a324335343e60278d4535c91157ba4
 PKG_SOURCE_DATE:=2024-03-29
 PKG_SOURCE_VERSION:=eb9bcb64185ac155c02cc1a604692c4b00368324
 PKG_ABI_VERSION:=$(call abi_version_str,$(PKG_SOURCE_DATE))
diff --git a/package/libs/udebug/Makefile b/package/libs/udebug/Makefile
index bfb13cf788..0a54515e74 100644
--- a/package/libs/udebug/Makefile
+++ b/package/libs/udebug/Makefile
@@ -11,7 +11,7 @@ PKG_NAME:=udebug
 CMAKE_INSTALL:=1
 PKG_SOURCE_PROTO:=git
 PKG_SOURCE_URL=$(PROJECT_GIT)/project/udebug.git
-PKG_MIRROR_HASH:=4dfe98b5d5813066dccd46d112502565b6961a4a7382083c529dfb423a626c2b
+PKG_MIRROR_HASH:=553a58a14b59dc7b22755557acbc74f0655a431442a4faca56d28cdb1ef14fb4
 PKG_SOURCE_DATE:=2023-12-06
 PKG_SOURCE_VERSION:=6d3f51f9fda706f0cf4732c762e4dbe8c21e12cf
 PKG_ABI_VERSION:=$(call abi_version_str,$(PKG_SOURCE_DATE))
diff --git a/package/network/config/netifd/Makefile b/package/network/config/netifd/Makefile
index a4400c9516..5edbb720cd 100644
--- a/package/network/config/netifd/Makefile
+++ b/package/network/config/netifd/Makefile
@@ -7,7 +7,7 @@ PKG_SOURCE_PROTO:=git
 PKG_SOURCE_URL=$(PROJECT_GIT)/project/netifd.git
 PKG_SOURCE_DATE:=2024-11-08
 PKG_SOURCE_VERSION:=34eb11eb6f5c62de480d6192f0ca840093127fee
-PKG_MIRROR_HASH:=c4c2bc626426d2cd73c9ece54b22ada6ed6254afa992f51b8725312aca240cfd
+PKG_MIRROR_HASH:=0886c58e63d8a2c90fb3394cc9d9054e3127d0b6896e6a32eabc0b3a846f4f67
 PKG_MAINTAINER:=Felix Fietkau <nbd@nbd.name>
 
 PKG_LICENSE:=GPL-2.0
diff --git a/package/network/utils/iwinfo/Makefile b/package/network/utils/iwinfo/Makefile
index d81ed62620..356ee5c7b4 100644
--- a/package/network/utils/iwinfo/Makefile
+++ b/package/network/utils/iwinfo/Makefile
@@ -13,7 +13,7 @@ PKG_SOURCE_PROTO:=git
 PKG_SOURCE_URL=$(PROJECT_GIT)/project/iwinfo.git
 PKG_SOURCE_DATE:=2024-10-20
 PKG_SOURCE_VERSION:=b94f066e3f5839b8509483cdd8f4f582a45fa233
-PKG_MIRROR_HASH:=ee0bce167707fe78f68a951b7ee1e0e61a92cae281e0e24eb709857ae849777e
+PKG_MIRROR_HASH:=a9007ae6718659643f25a0571dc5fea24f8357dc980c1ff1e332008a481e6e10
 PKG_MAINTAINER:=Jo-Philipp Wich <jo@mein.io>
 PKG_LICENSE:=GPL-2.0
 
diff --git a/package/system/procd/Makefile b/package/system/procd/Makefile
index 81e133f55b..97eb8b2194 100644
--- a/package/system/procd/Makefile
+++ b/package/system/procd/Makefile
@@ -12,7 +12,7 @@ PKG_RELEASE:=1
 
 PKG_SOURCE_PROTO:=git
 PKG_SOURCE_URL=$(PROJECT_GIT)/project/procd.git
-PKG_MIRROR_HASH:=56c5f71da3f68036c63ae59d01992785e74027726da5973297895985cd27c215
+PKG_MIRROR_HASH:=144c61f416738ab7e0543a37a718efe5727f34338ddaccb5064f5d1331be8869
 PKG_SOURCE_DATE:=2024-11-13
 PKG_SOURCE_VERSION:=7330fa55c5211eb7b3c675d1c7b8281b69b53553
 CMAKE_INSTALL:=1
diff --git a/package/system/rpcd/Makefile b/package/system/rpcd/Makefile
index da031db849..08639cc2c6 100644
--- a/package/system/rpcd/Makefile
+++ b/package/system/rpcd/Makefile
@@ -12,7 +12,7 @@ PKG_RELEASE:=1
 
 PKG_SOURCE_PROTO:=git
 PKG_SOURCE_URL=$(PROJECT_GIT)/project/rpcd.git
-PKG_MIRROR_HASH:=67b2cb985d8712a3e5a17ebf8c74fd35d553c8f9a4197616f9a3649a8740cc33
+PKG_MIRROR_HASH:=d3c6086672012effa4f06d04135ac52d1c4305cdb9242d972e55623e3614eab3
 PKG_SOURCE_DATE:=2024-09-17
 PKG_SOURCE_VERSION:=9f4b86e70352ab9ca6aa272d096419acc53e2390
 PKG_MAINTAINER:=Jo-Philipp Wich <jo@mein.io>
diff --git a/package/system/ubox/Makefile b/package/system/ubox/Makefile
index cb57e42e66..e54b9fa2d5 100644
--- a/package/system/ubox/Makefile
+++ b/package/system/ubox/Makefile
@@ -7,7 +7,7 @@ PKG_SOURCE_PROTO:=git
 PKG_SOURCE_URL=$(PROJECT_GIT)/project/ubox.git
 PKG_SOURCE_DATE:=2024-04-26
 PKG_SOURCE_VERSION:=85f1053019caf4cd333795760950235ee4529ba7
-PKG_MIRROR_HASH:=9e3fb6ab94854405fb91626a673b0547a061582c552ce719691be1bc8818da6c
+PKG_MIRROR_HASH:=0fea1a29aefc4b5932d739b0c942c182aef7a1ac44f467de80183b21dd739332
 CMAKE_INSTALL:=1
 
 PKG_LICENSE:=GPL-2.0
diff --git a/package/system/ubus/Makefile b/package/system/ubus/Makefile
index bef5e622d7..40dafd68f6 100644
--- a/package/system/ubus/Makefile
+++ b/package/system/ubus/Makefile
@@ -7,7 +7,7 @@ PKG_SOURCE_PROTO:=git
 PKG_SOURCE_URL=$(PROJECT_GIT)/project/ubus.git
 PKG_SOURCE_DATE:=2024-10-20
 PKG_SOURCE_VERSION:=252a9b0c1774790fb9c25735d5a09c27dba895db
-PKG_MIRROR_HASH:=72c21f02710d2314447670f1f1ea1833d2961f65fea3f9f94c060dda7c9d5914
+PKG_MIRROR_HASH:=475d96cc267370eb2e6ec808fc88285267cece1c5956d3966a060932d8b95175
 PKG_ABI_VERSION:=$(call abi_version_str,$(PKG_SOURCE_DATE))
 CMAKE_INSTALL:=1
 
diff --git a/package/utils/ucode/Makefile b/package/utils/ucode/Makefile
index 6ee0cf2a1d..8bee9b26e2 100644
--- a/package/utils/ucode/Makefile
+++ b/package/utils/ucode/Makefile
@@ -14,7 +14,7 @@ PKG_SOURCE_PROTO:=git
 PKG_SOURCE_URL=https://github.com/jow-/ucode.git
 PKG_SOURCE_DATE:=2024-07-22
 PKG_SOURCE_VERSION:=b610860dd4a0591ff586dd71a50f556a0ddafced
-PKG_MIRROR_HASH:=a5ec51dd989174422d3b19b022ff4f863d57eb559c9f08d54c0d10651f598357
+PKG_MIRROR_HASH:=eaad0b7f67bcf6742596fad909c95e5bd1e4548dfde72514d23ca450310c8c56
 PKG_MAINTAINER:=Jo-Philipp Wich <jo@mein.io>
 PKG_LICENSE:=ISC
 

