From 31b43a255c2fc56b6cd6d6ccd70a98346d8df78d Mon Sep 17 00:00:00 2001
From: Bo Jiao <Bo.Jiao@mediatek.com>
Date: Fri, 1 Apr 2022 09:15:21 +0800
Subject: [PATCH 19/89] mtk: mac80211: it's invalid case when frag_threshold is
 greater than 2346

Signed-off-by: Bo Jiao <Bo.Jiao@mediatek.com>
---
 net/wireless/nl80211.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/net/wireless/nl80211.c b/net/wireless/nl80211.c
index 93b37dd..abda8ca 100644
--- a/net/wireless/nl80211.c
+++ b/net/wireless/nl80211.c
@@ -3832,6 +3832,9 @@ static int nl80211_set_wiphy(struct sk_buff *skb, struct genl_info *info)
 			goto out;
 		}
 
+		if (frag_threshold >= 2346)
+			frag_threshold = (u32) -1;
+
 		if (frag_threshold != (u32) -1) {
 			/*
 			 * Fragments (apart from the last one) are required to
-- 
2.18.0

