From baa27bc8d70bab050630a127c7d58d140ac9ed21 Mon Sep 17 00:00:00 2001
From: ye he <ye.he@mediatek.com>
Date: Wed, 22 Feb 2023 16:09:32 +0800
Subject: [PATCH 27/61] mtk: mac80211: add send bar action when recieve addba
 rsp

Signed-off-by: ye he <ye.he@mediatek.com>
---
 net/mac80211/agg-tx.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/net/mac80211/agg-tx.c b/net/mac80211/agg-tx.c
index 5cf478e..8480b64 100644
--- a/net/mac80211/agg-tx.c
+++ b/net/mac80211/agg-tx.c
@@ -1080,7 +1080,8 @@ next:
 
 		tid_tx->buf_size = buf_size;
 		tid_tx->amsdu = amsdu;
-
+		ieee80211_send_bar(&sta->sdata->vif, sta->sta.addr,
+					   tid, 0);
 		if (test_bit(HT_AGG_STATE_DRV_READY, &tid_tx->state))
 			ieee80211_agg_tx_operational(local, sta, tid);
 
-- 
2.39.2

