From 1b24a353f3304b569dd17d28f9633fbc213651b2 Mon Sep 17 00:00:00 2001
From: "himanshu.goyal" <himanshu.goyal@mediatek.com>
Date: Tue, 2 May 2023 15:08:42 +0800
Subject: [PATCH] wifi: mt76: mt7915: Add 5G UNII4 support.

---
 mac80211.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/mac80211.c b/mac80211.c
index 4c88710..8ac1448 100644
--- a/mac80211.c
+++ b/mac80211.c
@@ -82,6 +82,7 @@ static const struct ieee80211_channel mt76_channels_5ghz[] = {
 	CHAN5G(165, 5825),
 	CHAN5G(169, 5845),
 	CHAN5G(173, 5865),
+	CHAN5G(177, 5885),
 
 	CHAN5G(184, 4920),
 	CHAN5G(188, 4940),
-- 
2.18.0

