From 2d4b5900f533b412215b14125d766c84f43e5bdb Mon Sep 17 00:00:00 2001
From: Sujuan Chen <sujuan.chen@mediatek.com>
Date: Tue, 27 Dec 2022 09:45:14 +0800
Subject: [PATCH] mt76: mt7915: add mt7915_net_setup_tc to support wifi2wifi
 offload

Signed-off-by: Sujuan Chen <sujuan.chen@mediatek.com>
---
 mt7915/main.c | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/mt7915/main.c b/mt7915/main.c
index 28509d94..ad5cd002 100644
--- a/mt7915/main.c
+++ b/mt7915/main.c
@@ -1598,6 +1598,21 @@ mt7915_net_fill_receive_path(struct ieee80211_hw *hw,
 	return 0;
 }
 
+static int mt7915_net_setup_tc(struct ieee80211_hw *hw,
+			      struct net_device *ndev,
+			      int type, void *type_data)
+{
+	struct mt7915_dev *dev = mt7915_hw_dev(hw);
+	struct mtk_wed_device *wed = &dev->mt76.mmio.wed;
+
+	if (!mtk_wed_device_active(wed))
+		return -ENODEV;
+
+	mtk_wed_device_setup_tc(wed, ndev, type, type_data);
+
+	return 0;
+}
+
 #endif
 
 const struct ieee80211_ops mt7915_ops = {
@@ -1651,5 +1666,6 @@ const struct ieee80211_ops mt7915_ops = {
 #ifdef CONFIG_NET_MEDIATEK_SOC_WED
 	.net_fill_forward_path = mt7915_net_fill_forward_path,
 	.net_fill_receive_path = mt7915_net_fill_receive_path,
+	.net_setup_tc = mt7915_net_setup_tc,
 #endif
 };
-- 
2.18.0

