From 0190aaa6ea5792a58b511701d7e8a9ce123878ad Mon Sep 17 00:00:00 2001
From: Sujuan Chen <sujuan.chen@mediatek.com>
Date: Tue, 3 Jan 2023 17:47:46 +0800
Subject: [PATCH] mt76: mt7915: drop the wrong judgment condition to support
 ppdu txs

Signed-off-by: Sujuan Chen <sujuan.chen@mediatek.com>
---
 mt7915/mac.c | 3 ---
 1 file changed, 3 deletions(-)

diff --git a/mt7915/mac.c b/mt7915/mac.c
index d759912..1cdca15 100644
--- a/mt7915/mac.c
+++ b/mt7915/mac.c
@@ -1126,9 +1126,6 @@ static void mt7915_mac_add_txs(struct mt7915_dev *dev, void *data)
 	u16 wcidx;
 	u8 pid;
 
-	if (le32_get_bits(txs_data[0], MT_TXS0_TXS_FORMAT) > 1)
-		return;
-
 	wcidx = le32_get_bits(txs_data[2], MT_TXS2_WCID);
 	pid = le32_get_bits(txs_data[3], MT_TXS3_PID);
 
-- 
2.18.0

