From 3d53aa889d0ed418ac2743e98925f44a2efbf4e4 Mon Sep 17 00:00:00 2001
From: Peter Chiu <chui-hao.chiu@mediatek.com>
Date: Tue, 18 Jan 2022 20:29:44 +0800
Subject: [PATCH 02/15] mac80211: mtk: do not setup twt when twt responder is
 false

---
 net/mac80211/rx.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/net/mac80211/rx.c b/net/mac80211/rx.c
index cc11ac5..9f2a13b 100644
--- a/net/mac80211/rx.c
+++ b/net/mac80211/rx.c
@@ -3298,6 +3298,9 @@ ieee80211_process_rx_twt_action(struct ieee80211_rx_data *rx)
 	if (sdata->vif.type != NL80211_IFTYPE_AP)
 		return false;
 
+	if (!sdata->vif.bss_conf.twt_responder)
+		return false;
+
 	if (!rx->local->ops->add_twt_setup)
 		return false;
 
-- 
2.25.1

