Index: drivers/net/phy/Kconfig
===================================================================
--- a/drivers/net/phy/Kconfig
+++ b/drivers/net/phy/Kconfig
@@ -350,6 +350,11 @@ config AIROHA_EN8801S_PHY
 	depends on HWMON || HWMON=n
 	select MDIO_I2C
 
+config AIROHA_EN8801SC_PHY
+	tristate "Drivers for Airoha EN8801S Gigabit PHYs for MediaTek SoC."
+	---help---
+	  Currently supports the Airoha EN8801S PHY for MediaTek SoC.
+
 config AIROHA_EN8811H_PHY
 	tristate "Drivers for Airoha EN8811H 2.5G Gigabit PHY"
 	---help---
Index: drivers/net/phy/Makefile
===================================================================
--- a/drivers/net/phy/Makefile
+++ b/drivers/net/phy/Makefile
@@ -68,5 +68,6 @@ ifdef CONFIG_HWMON
 aquantia-objs			+= aquantia_hwmon.o
 endif
+obj-$(CONFIG_AIROHA_EN8801SC_PHY)	+= en8801sc.o
 obj-$(CONFIG_AIROHA_EN8811H_PHY)	+= air_en8811h.o
 obj-$(CONFIG_AQUANTIA_PHY)	+= aquantia.o
 obj-$(CONFIG_AX88796B_PHY)	+= ax88796b.o
