--- a/tools/mkimage/Makefile
+++ b/tools/mkimage/Makefile
@@ -43,11 +43,13 @@ endef
 define Host/Install
 	$(CP) $(HOST_BUILD_DIR)/tools/mkimage $(STAGING_DIR_HOST)/bin/
 	$(CP) $(HOST_BUILD_DIR)/tools/mkenvimage $(STAGING_DIR_HOST)/bin/
+	$(CP) $(HOST_BUILD_DIR)/tools/fdt_add_pubkey $(STAGING_DIR_HOST)/bin/
 endef
 
 define Host/Clean
 	rm -f $(STAGING_DIR_HOST)/bin/mkimage
 	rm -f $(STAGING_DIR_HOST)/bin/mkenvimage
+	rm -f $(STAGING_DIR_HOST)/bin/fdt_add_pubkey
 endef
 
 $(eval $(call HostBuild))
