--- a/target/linux/mediatek/image/filogic.mk
+++ b/target/linux/mediatek/image/filogic.mk
@@ -1902,10 +1902,12 @@ define Device/mediatek_mt7988a-rfb
 	mt7988a-rfb-eth1-i2p5g-phy \
 	mt7988a-rfb-eth1-mxl \
 	mt7988a-rfb-eth1-sfp \
+	mt7988a-rfb-eth1-sfp-an8831x \
 	mt7988a-rfb-eth2-aqr \
 	mt7988a-rfb-eth2-an8831x \
 	mt7988a-rfb-eth2-cux3410 \
 	mt7988a-rfb-eth2-mxl \
+	mt7988a-rfb-eth2-mxl86252-an8831x \
 	mt7988a-rfb-eth2-sfp \
 	mt7988a-rfb-spidev \
 	mt7988a-rfb-4pcie \
