--- a/target/linux/mediatek/image/filogic.mk
+++ b/target/linux/mediatek/image/filogic.mk
@@ -1873,6 +1873,7 @@ define Device/mediatek_mt7986a-rfb
   DEVICE_DTS_OVERLAY:= \
 	mt7986-spim-nand \
 	mt7986-spim-nand-nmbm \
+	mt7986-spim-nor \
 	mt7986-spidev
   DEVICE_DTS_DIR := $(DTS_DIR)/
   DEVICE_DTC_FLAGS := --pad 4096
@@ -1898,6 +1899,7 @@ define Device/mediatek_mt7986b-rfb
   DEVICE_DTS_OVERLAY:= \
 	mt7986-spim-nand \
 	mt7986-spim-nand-nmbm \
+	mt7986-spim-nor \
 	mt7986-spidev
   DEVICE_DTS_DIR := $(DTS_DIR)/
   DEVICE_DTC_FLAGS := --pad 4096
