--- a/target/linux/mediatek/image/filogic.mk
+++ b/target/linux/mediatek/image/filogic.mk
@@ -1829,7 +1829,8 @@ define Device/mediatek_mt7986a-rfb
   DEVICE_DTS := mt7986a-rfb
   DEVICE_DTS_OVERLAY:= \
 	mt7986-spim-nand \
-	mt7986-spim-nand-nmbm
+	mt7986-spim-nand-nmbm \
+	mt7986-spidev
   DEVICE_DTS_DIR := $(DTS_DIR)/
   DEVICE_DTC_FLAGS := --pad 4096
   DEVICE_DTS_LOADADDR := 0x43f00000
@@ -1853,7 +1854,8 @@ define Device/mediatek_mt7986b-rfb
   DEVICE_DTS := mt7986b-rfb
   DEVICE_DTS_OVERLAY:= \
 	mt7986-spim-nand \
-	mt7986-spim-nand-nmbm
+	mt7986-spim-nand-nmbm \
+	mt7986-spidev
   DEVICE_DTS_DIR := $(DTS_DIR)/
   DEVICE_DTC_FLAGS := --pad 4096
   DEVICE_DTS_LOADADDR := 0x43f00000
