--- a/target/linux/mediatek/image/filogic.mk
+++ b/target/linux/mediatek/image/filogic.mk
@@ -1840,11 +1840,11 @@ TARGET_DEVICES += konka_komi-a31
 
 define Device/mediatek_mt7981-rfb
   DEVICE_VENDOR := MediaTek
-  DEVICE_MODEL := MT7981 rfb
+  DEVICE_MODEL := MT7981 rfb (DT-overlay)
   DEVICE_DTS := mt7981-rfb
   DEVICE_DTS_OVERLAY:= \
 	mt7981-rfb-spim-nor \
-	mt7981-rfb-spim-nand \
+	mt7981-rfb-spim-nand-nmbm \
 	mt7981-rfb-mxl-2p5g-phy-eth1 \
 	mt7981-rfb-mxl-2p5g-phy-swp5
   DEVICE_DTS_DIR := $(DTS_DIR)/
