diff --git a/target/linux/mediatek/image/filogic.mk b/target/linux/mediatek/image/filogic.mk
index ae5e085..79895b2 100644
--- a/target/linux/mediatek/image/filogic.mk
+++ b/target/linux/mediatek/image/filogic.mk
@@ -1871,6 +1872,7 @@ define Device/mediatek_mt7986a-rfb
   DEVICE_MODEL := MT7986A rfb (DT-overlay)
   DEVICE_DTS := mt7986a-rfb
   DEVICE_DTS_OVERLAY:= \
+	mt7986-sfp-eth1 \
 	mt7986-spim-nand \
 	mt7986-spim-nand-nmbm \
 	mt7986-spidev
@@ -1896,6 +1898,7 @@ define Device/mediatek_mt7986b-rfb
   DEVICE_MODEL := MT7986B rfb (DT-overlay)
   DEVICE_DTS := mt7986b-rfb
   DEVICE_DTS_OVERLAY:= \
+	mt7986-sfp-eth1 \
 	mt7986-spim-nand \
 	mt7986-spim-nand-nmbm \
 	mt7986-spidev
