--- a/boot/image-cipher.c
+++ b/boot/image-cipher.c
@@ -49,6 +49,28 @@ struct cipher_algo cipher_algos[] = {
 		.encrypt = image_aes_encrypt,
 		.decrypt = image_aes_decrypt,
 		.add_cipher_data = image_aes_add_cipher_data
+	},
+	{
+		.name = "tee_aes256",
+		.key_len = AES256_KEY_LENGTH,
+		.iv_len  = AES_BLOCK_LENGTH,
+#if IMAGE_ENABLE_ENCRYPT
+		.calculate_type = EVP_aes_256_cbc,
+#endif
+		.encrypt = image_aes_encrypt,
+		.decrypt = image_aes_decrypt,
+		.add_cipher_data = image_aes_add_cipher_data
+	},
+	{
+		.name = "optee_aes256",
+		.key_len = AES256_KEY_LENGTH,
+		.iv_len  = AES_BLOCK_LENGTH,
+#if IMAGE_ENABLE_ENCRYPT
+		.calculate_type = EVP_aes_256_cbc,
+#endif
+		.encrypt = image_aes_encrypt,
+		.decrypt = image_aes_decrypt,
+		.add_cipher_data = image_aes_add_cipher_data
 	}
 };
 
