--- a/ax_config.h
+++ b/ax_config.h
@@ -22,11 +22,11 @@
 #define ENABLE_COE                  0
 #define ENABLE_LSO                  1
 
-#define ENABLE_PTP_FUNC             0
+#define ENABLE_PTP_FUNC             1
 #define ENABLE_PTP_DEBUG            0
 
 #define ENABLE_PTP_125M_CLK         0
-#define ENABLE_PTP_PPS              0
+#define ENABLE_PTP_PPS              1
 
 #define ENABLE_PTP_NORMAL_PKT       0
 
