From 62aafa5960541ea5a9d8ad50bce4e7e5940f4486 Mon Sep 17 00:00:00 2001
From: Mason Chang <mason-cw.chang@mediatek.com>
Date: Thu, 18 Sep 2025 16:43:38 +0800
Subject: [PATCH] thermal: mediatek: backport kernel-6.17 lvts driver

---
 drivers/thermal/mediatek/lvts_thermal.c | 56 +++++++++++++++++--------
 1 file changed, 39 insertions(+), 17 deletions(-)

--- a/drivers/thermal/mediatek/lvts_thermal.c
+++ b/drivers/thermal/mediatek/lvts_thermal.c
@@ -453,7 +453,7 @@ static irqreturn_t lvts_ctrl_irq_handler
 {
 	irqreturn_t iret = IRQ_NONE;
 	u32 value;
-	u32 masks[] = {
+	static const u32 masks[] = {
 		LVTS_INT_SENSOR0,
 		LVTS_INT_SENSOR1,
 		LVTS_INT_SENSOR2,
@@ -582,7 +582,7 @@ static irqreturn_t lvts_irq_handler(int
 	return iret;
 }
 
-static struct thermal_zone_device_ops lvts_ops = {
+static const struct thermal_zone_device_ops lvts_ops = {
 	.get_temp = lvts_get_temp,
 	.set_trips = lvts_set_trips,
 };
@@ -1931,7 +1931,7 @@ static const struct dev_pm_ops lvts_pm_o
 
 static struct platform_driver lvts_driver = {
 	.probe = lvts_probe,
-	.remove_new = lvts_remove,
+	.remove = lvts_remove,
 	.driver = {
 		.name = "mtk-lvts-thermal",
 		.of_match_table = lvts_of_match,
