--- a/drivers/clk/mediatek/clk-mt7987-infracfg.c
+++ b/drivers/clk/mediatek/clk-mt7987-infracfg.c
@@ -25,15 +25,12 @@
 static DEFINE_SPINLOCK(mt7987_clk_lock);
 
 static const char *const infra_mux_uart0_parents[] = { "csw_infra_f26m_sel",
-						       "infra_hf_66m_uart0_pck",
 						       "uart_sel" };
 
 static const char *const infra_mux_uart1_parents[] = { "csw_infra_f26m_sel",
-						       "infra_hf_66m_uart1_pck",
 						       "uart_sel" };
 
 static const char *const infra_mux_uart2_parents[] = { "csw_infra_f26m_sel",
-						       "infra_hf_66m_uart1_pck",
 						       "uart_sel" };
 
 static const char *const infra_mux_spi0_parents[] = {
