From 59c4b77798c213a1766b3dac36a9de08145f063e Mon Sep 17 00:00:00 2001
From: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
Date: Thu, 9 Mar 2023 18:45:04 +0800
Subject: [PATCH 21/22] mt76: testmode: add chainmask hacking for eagle band 2
 4T5R

Signed-off-by: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
---
 mt7996/testmode.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/mt7996/testmode.c b/mt7996/testmode.c
index 6d7cdbd5..0d4d9138 100644
--- a/mt7996/testmode.c
+++ b/mt7996/testmode.c
@@ -463,6 +463,7 @@ mt7996_tm_set_params(struct mt76_phy *mphy, struct nlattr **tb,
 		return 0;
 
 	chainmask = chainmask >> dev->chainshift[band_idx];
+	chainmask = 0x1f;	/* hacking for eagle band2 4T5R */
 	if (td->tx_antenna_mask > chainmask)
 		return -EINVAL;
 
-- 
2.39.2

