From 7251f2a0f7d1f5eda27611ff658328c2fe8edbf1 Mon Sep 17 00:00:00 2001
From: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
Date: Wed, 27 Mar 2024 17:50:16 +0800
Subject: [PATCH 048/223] mtk: mt76: mt7996: add kite part number support

Change-Id: Ib06648398f18b47c324e18b476a57444d929608f
Signed-off-by: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
---
 mt7996/eeprom.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/mt7996/eeprom.c b/mt7996/eeprom.c
index 3ec146d0..f9bd4d91 100644
--- a/mt7996/eeprom.c
+++ b/mt7996/eeprom.c
@@ -372,7 +372,6 @@ static int mt7996_eeprom_parse_efuse_hw_cap(struct mt7996_phy *phy,
 	if (ret)
 		return ret;
 
-	cap = 0x4b249248;	/* internal hardcode */
 	if (cap) {
 		u8 band_offs = phy->mt76->band_idx * 3;
 
-- 
2.45.2

