From 8bb932b534480f0025fdbd57270b194ed7738e2d Mon Sep 17 00:00:00 2001
From: Shayne Chen <shayne.chen@mediatek.com>
Date: Wed, 15 May 2024 11:10:17 +0800
Subject: [PATCH 133/199] mtk: mt76: mt7996: fix stop_tx_ba_session warning

---
 mt7996/main.c | 11 +++++++----
 mt7996/mcu.c  |  6 ++++++
 2 files changed, 13 insertions(+), 4 deletions(-)

diff --git a/mt7996/main.c b/mt7996/main.c
index 0a87cfad..c7ec01f2 100644
--- a/mt7996/main.c
+++ b/mt7996/main.c
@@ -1079,7 +1079,8 @@ static void mt7996_remove_link_sta(struct mt7996_dev *dev,
 				   struct ieee80211_bss_conf *conf,
 				   struct mt7996_bss_conf *mconf,
 				   struct ieee80211_link_sta *link_sta,
-				   struct mt7996_link_sta *mlink)
+				   struct mt7996_link_sta *mlink,
+				   bool last_link)
 {
 	struct ieee80211_sta *sta = link_sta->sta;
 	struct mt7996_sta *msta = (struct mt7996_sta *)sta->drv_priv;
@@ -1091,7 +1092,7 @@ static void mt7996_remove_link_sta(struct mt7996_dev *dev,
 	for (i = 0; i < ARRAY_SIZE(mlink->wcid.aggr); i++)
 			mt76_rx_aggr_stop(&dev->mt76, &mlink->wcid, i);
 
-	if (sta->mlo)
+	if (sta->mlo && last_link)
 		mt7996_mcu_teardown_mld_sta(dev, mconf, mlink);
 	else
 		mt7996_mcu_add_sta(dev, conf, mconf, link_sta, mlink, false, false);
@@ -1205,7 +1206,8 @@ static int mt7996_add_link_sta(struct mt7996_dev *dev,
 
 	return 0;
 error:
-	mt7996_remove_link_sta(dev, conf, mconf, link_sta, mlink);
+	mt7996_remove_link_sta(dev, conf, mconf, link_sta, mlink,
+			       hweight16(sta->valid_links) <= 1);
 	return ret;
 }
 
@@ -1230,8 +1232,9 @@ mt7996_mac_sta_remove_links(struct mt7996_dev *dev, struct ieee80211_vif *vif,
 			link_conf_dereference_protected(vif, link_id);
 		struct ieee80211_link_sta *link_sta =
 			link_sta_dereference_protected(sta, link_id);
+		bool last_link = rem == sta->valid_links && link_id == __fls(rem);
 
-		mt7996_remove_link_sta(dev, conf, mconf, link_sta, mlink);
+		mt7996_remove_link_sta(dev, conf, mconf, link_sta, mlink, last_link);
 	}
 }
 
diff --git a/mt7996/mcu.c b/mt7996/mcu.c
index 0a0df745..48ff11d9 100644
--- a/mt7996/mcu.c
+++ b/mt7996/mcu.c
@@ -1735,6 +1735,9 @@ int mt7996_mcu_add_tx_ba(struct mt7996_dev *dev,
 			mconf_dereference_protected(msta->vif, link_id);
 		int ret;
 
+		if (!mlink || !mconf)
+			continue;
+
 		if (enable && !params->amsdu)
 			mlink->wcid.amsdu = false;
 
@@ -1769,6 +1772,9 @@ int mt7996_mcu_add_rx_ba(struct mt7996_dev *dev,
 			mconf_dereference_protected(msta->vif, link_id);
 		int ret;
 
+		if (!mlink || !mconf)
+			continue;
+
 		ret = mt7996_mcu_sta_ba(dev, &mconf->mt76, params, &mlink->wcid,
 					enable, false);
 		if (ret)
-- 
2.18.0

