From ce237c34f5eba3f1d90035b885a89579284a930b Mon Sep 17 00:00:00 2001
From: Peter Chiu <chui-hao.chiu@mediatek.com>
Date: Sat, 28 Sep 2024 11:19:18 +0800
Subject: [PATCH 233/265] mtk: mt76: mt7990: add DMA ring for TxFree

Signed-off-by: Peter Chiu <chui-hao.chiu@mediatek.com>
---
 mt7996/dma.c    | 39 +++++++++++++++++++++++++++++++++++++++
 mt7996/mt7996.h |  2 ++
 mt7996/regs.h   |  2 ++
 3 files changed, 43 insertions(+)

diff --git a/mt7996/dma.c b/mt7996/dma.c
index f3a9ed8c..5c2e6422 100644
--- a/mt7996/dma.c
+++ b/mt7996/dma.c
@@ -75,6 +75,11 @@ static void mt7996_dma_config(struct mt7996_dev *dev)
 		break;
 	case MT7990_DEVICE_ID:
 		RXQ_CONFIG(MT_RXQ_BAND1, WFDMA0, MT_INT_RX_DONE_BAND1, MT7996_RXQ_BAND1);
+		RXQ_CONFIG(MT_RXQ_TXFREE_BAND0, WFDMA0,
+			   MT_INT_RX_TXFREE_BAND0_MT7990, MT7990_RXQ_TXFREE0);
+		if (dev->hif2)
+			RXQ_CONFIG(MT_RXQ_TXFREE_BAND1, WFDMA0,
+				   MT_INT_RX_TXFREE_BAND1_MT7990, MT7990_RXQ_TXFREE1);
 		break;
 	case MT7996_DEVICE_ID:
 	default:
@@ -178,6 +183,12 @@ static void __mt7996_dma_prefetch(struct mt7996_dev *dev, u32 ofs)
 
 		queue = is_mt7996(&dev->mt76) ? MT_RXQ_BAND2_WA : MT_RXQ_BAND1_WA;
 		mt76_wr(dev, MT_RXQ_BAND1_CTRL(queue) + ofs, PREFETCH(0x2));
+	} else {
+		mt76_wr(dev, MT_RXQ_BAND1_CTRL(MT_RXQ_TXFREE_BAND0) + ofs,
+			PREFETCH(0x8));
+		if (dev->hif2)
+			mt76_wr(dev, MT_RXQ_BAND1_CTRL(MT_RXQ_TXFREE_BAND1) + ofs,
+				PREFETCH(0x8));
 	}
 
 	mt76_wr(dev, MT_RXQ_BAND1_CTRL(MT_RXQ_MAIN) + ofs, PREFETCH(0x10));
@@ -670,6 +681,34 @@ int mt7996_dma_init(struct mt7996_dev *dev)
 				       MT_RXQ_RING_BASE(MT_RXQ_MAIN_WA));
 		if (ret)
 			return ret;
+	} else {
+		/* Todo: support hw path */
+		/* if (mtk_wed_device_active(wed)) { */
+		/* 	dev->mt76.q_rx[MT_RXQ_TXFREE_BAND0].flags = MT_WED_Q_TXFREE; */
+		/* 	dev->mt76.q_rx[MT_RXQ_TXFREE_BAND0].wed = wed; */
+		/* } */
+		ret = mt76_queue_alloc(dev, &dev->mt76.q_rx[MT_RXQ_TXFREE_BAND0],
+				       MT_RXQ_ID(MT_RXQ_TXFREE_BAND0),
+				       MT7996_RX_MCU_RING_SIZE,
+				       MT7996_RX_BUF_SIZE,
+				       MT_RXQ_RING_BASE(MT_RXQ_TXFREE_BAND0));
+		if (ret)
+			return ret;
+	}
+
+	if (!mt7996_has_wa(dev) && dev->hif2) {
+		/* Todo: support hw path */
+		/* if (mtk_wed_device_active(wed)) { */
+		/* 	dev->mt76.q_rx[MT_RXQ_TXFREE_BAND1].flags = MT_WED_Q_TXFREE; */
+		/* 	dev->mt76.q_rx[MT_RXQ_TXFREE_BAND1].wed = wed; */
+		/* } */
+		ret = mt76_queue_alloc(dev, &dev->mt76.q_rx[MT_RXQ_TXFREE_BAND1],
+				       MT_RXQ_ID(MT_RXQ_TXFREE_BAND1),
+				       MT7996_RX_MCU_RING_SIZE,
+				       MT7996_RX_BUF_SIZE,
+				       MT_RXQ_RING_BASE(MT_RXQ_TXFREE_BAND1));
+		if (ret)
+			return ret;
 	}
 
 	if (mt7996_band_valid(dev, MT_BAND2)) {
diff --git a/mt7996/mt7996.h b/mt7996/mt7996.h
index 722aa5cc..0fac7693 100644
--- a/mt7996/mt7996.h
+++ b/mt7996/mt7996.h
@@ -260,6 +260,8 @@ enum mt7996_rxq_id {
 	MT7996_RXQ_TXFREE1 = 9,
 	MT7996_RXQ_TXFREE2 = 7,
 	MT7996_RXQ_RRO_IND = 0,
+	MT7990_RXQ_TXFREE0 = 6,
+	MT7990_RXQ_TXFREE1 = 7,
 };
 
 enum {
diff --git a/mt7996/regs.h b/mt7996/regs.h
index c198b136..e74b1c85 100644
--- a/mt7996/regs.h
+++ b/mt7996/regs.h
@@ -532,6 +532,8 @@ enum offs_rev {
 #define MT_INT_RX_TXFREE_BAND1			BIT(15)
 #define MT_INT_RX_TXFREE_TRI			BIT(15)
 #define MT_INT_RX_TXFREE_BAND1_EXT		BIT(19) /* for mt7992 two PCIE*/
+#define MT_INT_RX_TXFREE_BAND0_MT7990		BIT(14)
+#define MT_INT_RX_TXFREE_BAND1_MT7990		BIT(15)
 #define MT_INT_RX_DONE_BAND2_EXT		BIT(23)
 #define MT_INT_RX_TXFREE_EXT			BIT(26)
 #define MT_INT_MCU_CMD				BIT(29)
-- 
2.45.2

