From 948c361fbe90e19b1c1c46e8e1ad7b9c7ebdd7f2 Mon Sep 17 00:00:00 2001
From: Peter Chiu <chui-hao.chiu@mediatek.com>
Date: Sat, 28 Sep 2024 13:05:56 +0800
Subject: [PATCH 234/265] mtk: mt76: mt7990: modify wtbl CRs for MT7990

Signed-off-by: Peter Chiu <chui-hao.chiu@mediatek.com>
---
 mt7996/mmio.c | 15 +++++++++++++++
 mt7996/regs.h | 15 ++++++++++-----
 2 files changed, 25 insertions(+), 5 deletions(-)

diff --git a/mt7996/mmio.c b/mt7996/mmio.c
index 7f966f38..2e15fe3d 100644
--- a/mt7996/mmio.c
+++ b/mt7996/mmio.c
@@ -60,6 +60,11 @@ static const u32 mt7996_offs[] = {
 	[HIF_REMAP_BASE_L2]	= 0x1000,
 	[CBTOP1_PHY_END]	= 0x77ffffff,
 	[INFRA_MCU_END]		= 0x7c3fffff,
+	[WTBLON_WDUCR]		= 0x370,
+	[WTBL_UPDATE]		= 0x380,
+	[WTBL_ITCR]		= 0x3b0,
+	[WTBL_ITCR0]		= 0x3b8,
+	[WTBL_ITCR1]		= 0x3bc,
 };
 
 static const u32 mt7992_offs[] = {
@@ -92,6 +97,11 @@ static const u32 mt7992_offs[] = {
 	[HIF_REMAP_BASE_L2]	= 0x1000,
 	[CBTOP1_PHY_END]	= 0x77ffffff,
 	[INFRA_MCU_END]		= 0x7c3fffff,
+	[WTBLON_WDUCR]		= 0x370,
+	[WTBL_UPDATE]		= 0x380,
+	[WTBL_ITCR]		= 0x3b0,
+	[WTBL_ITCR0]		= 0x3b8,
+	[WTBL_ITCR1]		= 0x3bc,
 };
 
 static const u32 mt7990_offs[] = {
@@ -124,6 +134,11 @@ static const u32 mt7990_offs[] = {
 	[HIF_REMAP_BASE_L2]	= 0x110000,
 	[CBTOP1_PHY_END]	= 0x7fffffff,
 	[INFRA_MCU_END]		= 0x7cffffff,
+	[WTBLON_WDUCR]		= 0x400,
+	[WTBL_UPDATE]		= 0x410,
+	[WTBL_ITCR]		= 0x440,
+	[WTBL_ITCR0]		= 0x448,
+	[WTBL_ITCR1]		= 0x44c,
 };
 
 static const struct __map mt7996_reg_map[] = {
diff --git a/mt7996/regs.h b/mt7996/regs.h
index e74b1c85..6b0f0775 100644
--- a/mt7996/regs.h
+++ b/mt7996/regs.h
@@ -70,6 +70,11 @@ enum offs_rev {
 	HIF_REMAP_BASE_L2,
 	CBTOP1_PHY_END,
 	INFRA_MCU_END,
+	WTBLON_WDUCR,
+	WTBL_UPDATE,
+	WTBL_ITCR,
+	WTBL_ITCR0,
+	WTBL_ITCR1,
 	__MT_OFFS_MAX,
 };
 
@@ -306,19 +311,19 @@ enum offs_rev {
 /* WTBLON TOP */
 #define MT_WTBLON_TOP_BASE			0x820d4000
 #define MT_WTBLON_TOP(ofs)			(MT_WTBLON_TOP_BASE + (ofs))
-#define MT_WTBLON_TOP_WDUCR			MT_WTBLON_TOP(0x370)
+#define MT_WTBLON_TOP_WDUCR			MT_WTBLON_TOP(__OFFS(WTBLON_WDUCR))
 #define MT_WTBLON_TOP_WDUCR_GROUP		GENMASK(4, 0)
 
-#define MT_WTBL_UPDATE				MT_WTBLON_TOP(0x380)
+#define MT_WTBL_UPDATE				MT_WTBLON_TOP(__OFFS(WTBL_UPDATE))
 #define MT_WTBL_UPDATE_WLAN_IDX			GENMASK(11, 0)
 #define MT_WTBL_UPDATE_ADM_COUNT_CLEAR		BIT(14)
 #define MT_WTBL_UPDATE_BUSY			BIT(31)
 
-#define MT_WTBL_ITCR				MT_WTBLON_TOP(0x3b0)
+#define MT_WTBL_ITCR				MT_WTBLON_TOP(__OFFS(WTBL_ITCR))
 #define MT_WTBL_ITCR_WR				BIT(16)
 #define MT_WTBL_ITCR_EXEC			BIT(31)
-#define MT_WTBL_ITDR0				MT_WTBLON_TOP(0x3b8)
-#define MT_WTBL_ITDR1				MT_WTBLON_TOP(0x3bc)
+#define MT_WTBL_ITDR0				MT_WTBLON_TOP(__OFFS(WTBL_ITCR0))
+#define MT_WTBL_ITDR1				MT_WTBLON_TOP(__OFFS(WTBL_ITCR1))
 #define MT_WTBL_SPE_IDX_SEL			BIT(6)
 
 /* WTBL */
-- 
2.45.2

