From ba80aca90cdfbc98b8a21489f0fc89114c94f2d1 Mon Sep 17 00:00:00 2001
From: Peter Chiu <chui-hao.chiu@mediatek.com>
Date: Mon, 13 Feb 2023 11:03:53 +0800
Subject: [PATCH 025/126] mtk: hostapd: 6G band does not require DFS

---
 src/ap/dfs.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/ap/dfs.c b/src/ap/dfs.c
index 42cce2dce..86598a18a 100644
--- a/src/ap/dfs.c
+++ b/src/ap/dfs.c
@@ -1538,6 +1538,7 @@ int hostapd_is_dfs_required(struct hostapd_iface *iface)
 	if ((!(iface->drv_flags & WPA_DRIVER_FLAGS_DFS_OFFLOAD) &&
 	     !iface->conf->ieee80211h) ||
 	    !iface->current_mode ||
+	    is_6ghz_freq(iface->freq) ||
 	    iface->current_mode->mode != HOSTAPD_MODE_IEEE80211A)
 		return 0;
 
-- 
2.18.0

