From d29370b7ec78ba410d0877441206849e197b1609 Mon Sep 17 00:00:00 2001
From: mtk23510 <rudra.shahi@mediatek.com>
Date: Fri, 26 May 2023 14:52:35 +0800
Subject: [PATCH 042/126] mtk: hostapd: Add support for gtk rekeying in hostapd
 cli

Signed-off-by: mtk23510 <rudra.shahi@mediatek.com>
---
 hostapd/hostapd_cli.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/hostapd/hostapd_cli.c b/hostapd/hostapd_cli.c
index dba805e8a..e43d515d5 100644
--- a/hostapd/hostapd_cli.c
+++ b/hostapd/hostapd_cli.c
@@ -1309,6 +1309,15 @@ static int hostapd_cli_cmd_stop_ap(struct wpa_ctrl *ctrl, int argc,
 }
 
 
+#ifdef CONFIG_TESTING_OPTIONS
+static int hostapd_cli_cmd_rekey_gtk(struct wpa_ctrl *ctrl, int argc,
+				      char *argv[])
+{
+	return wpa_ctrl_command(ctrl, "REKEY_GTK");
+}
+#endif
+
+
 static int hostapd_cli_cmd_vendor(struct wpa_ctrl *ctrl, int argc, char *argv[])
 {
 	char cmd[256];
@@ -1861,6 +1870,10 @@ static const struct hostapd_cli_cmd hostapd_cli_commands[] = {
 	  "= update Beacon frame contents\n"},
 	{ "stop_ap", hostapd_cli_cmd_stop_ap, NULL,
 	  "= stop AP\n"},
+#ifdef CONFIG_TESTING_OPTIONS
+	{ "rekey_gtk", hostapd_cli_cmd_rekey_gtk, NULL,
+	  "= rekey gtk\n"},
+#endif
 	{ "erp_flush", hostapd_cli_cmd_erp_flush, NULL,
 	  "= drop all ERP keys"},
 	{ "log_level", hostapd_cli_cmd_log_level, NULL,
-- 
2.18.0

