From 97f988afe8673579444a9ae90c2db48a6d7527e7 Mon Sep 17 00:00:00 2001
From: Evelyn Tsai <evelyn.tsai@mediatek.com>
Date: Wed, 10 May 2023 13:11:34 +0800
Subject: [PATCH 049/126] mtk: hostapd: update eht operation element

---
 src/ap/ieee802_11_eht.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/src/ap/ieee802_11_eht.c b/src/ap/ieee802_11_eht.c
index b935ee889..a04eb23e0 100644
--- a/src/ap/ieee802_11_eht.c
+++ b/src/ap/ieee802_11_eht.c
@@ -237,9 +237,9 @@ u8 * hostapd_eid_eht_operation(struct hostapd_data *hapd, u8 *eid)
 
 	/* TODO: Fill in appropriate EHT-MCS max Nss information */
 	oper->basic_eht_mcs_nss_set[0] = 0x11;
-	oper->basic_eht_mcs_nss_set[1] = 0x00;
-	oper->basic_eht_mcs_nss_set[2] = 0x00;
-	oper->basic_eht_mcs_nss_set[3] = 0x00;
+	oper->basic_eht_mcs_nss_set[1] = 0x11;
+	oper->basic_eht_mcs_nss_set[2] = 0x11;
+	oper->basic_eht_mcs_nss_set[3] = 0x11;
 
 	if (!eht_oper_info_present)
 		return pos + elen;
-- 
2.18.0

