From 69f830f234d58e4bcbdafe70d28c366e87a65c66 Mon Sep 17 00:00:00 2001
From: Shayne Chen <shayne.chen@mediatek.com>
Date: Mon, 29 Jan 2024 11:24:28 +0800
Subject: [PATCH 060/126] mtk: hostapd: fix mld_assoc_link_id

Signed-off-by: Shayne Chen <shayne.chen@mediatek.com>
---
 src/ap/hostapd.c | 6 ++----
 1 file changed, 2 insertions(+), 4 deletions(-)

diff --git a/src/ap/hostapd.c b/src/ap/hostapd.c
index f64f17ee6..eefd4c5d3 100644
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -4111,11 +4111,9 @@ void hostapd_new_assoc_sta(struct hostapd_data *hapd, struct sta_info *sta,
 
 #ifdef CONFIG_IEEE80211BE
 	if (ap_sta_is_mld(hapd, sta)) {
-		if (sta->mld_assoc_link_id == hapd->mld_link_id) {
-			mld_assoc_link_id = sta->mld_assoc_link_id;
-		} else {
+		if (sta->mld_assoc_link_id != hapd->mld_link_id)
 			return;
-		}
+		mld_assoc_link_id = sta->mld_assoc_link_id;
 	}
 #endif /* CONFIG_IEEE80211BE */
         if (mld_assoc_link_id != -2)
-- 
2.18.0

