From d06dd0d45977ce098df718a29ffbc765896a2758 Mon Sep 17 00:00:00 2001
From: Shayne Chen <shayne.chen@mediatek.com>
Date: Mon, 29 Jan 2024 11:24:28 +0800
Subject: [PATCH 080/104] mtk: hostapd: fix mld_assoc_link_id

Signed-off-by: Shayne Chen <shayne.chen@mediatek.com>
---
 src/ap/hostapd.c | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/src/ap/hostapd.c b/src/ap/hostapd.c
index e4fc1f85a..f8b05de45 100644
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -4034,11 +4034,13 @@ void hostapd_new_assoc_sta(struct hostapd_data *hapd, struct sta_info *sta,
 	}
 
 #ifdef CONFIG_IEEE80211BE
-	if (ap_sta_is_mld(hapd, sta) &&
-	    sta->mld_assoc_link_id != hapd->mld_link_id)
-		return;
+	if (ap_sta_is_mld(hapd, sta)) {
+		if (sta->mld_assoc_link_id != hapd->mld_link_id)
+			return;
+		mld_assoc_link_id = sta->mld_assoc_link_id;
+	}
 #endif /* CONFIG_IEEE80211BE */
-        if (mld_assoc_link_id != -2)
+	if (mld_assoc_link_id != -2)
 		hostapd_prune_associations(hapd, sta->addr, mld_assoc_link_id);
 
 	ap_sta_clear_disconnect_timeouts(hapd, sta);
-- 
2.39.2

