--- a/package/libs/openssl/Makefile
+++ b/package/libs/openssl/Makefile
@@ -13,6 +13,7 @@ PKG_RELEASE:=1
 PKG_BUILD_FLAGS:=no-mips16 gc-sections no-lto
 
 PKG_BUILD_PARALLEL:=1
+PKG_BUILD_DEPENDS:=openssl/host
 
 PKG_BASE:=$(subst $(space),.,$(wordlist 1,2,$(subst .,$(space),$(PKG_VERSION))))
 PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.gz
@@ -61,6 +62,7 @@ PKG_CONFIG_DEPENDS:= \
 	CONFIG_OPENSSL_WITH_WHIRLPOOL
 
 include $(INCLUDE_DIR)/package.mk
+include $(INCLUDE_DIR)/host-build.mk
 include $(INCLUDE_DIR)/openssl-module.mk
 
 ifneq ($(CONFIG_CCACHE),)
@@ -428,6 +430,22 @@ define Package/openssl-util/install
 	$(INSTALL_BIN) $(PKG_INSTALL_DIR)/usr/bin/openssl $(1)/usr/bin/
 endef
 
+define Host/Configure
+	(cd $(HOST_BUILD_DIR); \
+		./Configure \
+			--prefix=$(STAGING_DIR_HOSTPKG)/ \
+			--libdir=lib \
+			--openssldir=/etc/ssl \
+			no-shared \
+			-static && \
+		{ [ -f $(STAMP_CONFIGURED) ] || make clean; } \
+	)
+endef
+
+define Host/Install
+	$(CP) $(HOST_BUILD_DIR)/apps/openssl $(STAGING_DIR_HOSTPKG)/bin/openssl3
+endef
+
 $(eval $(call BuildPackage,libopenssl))
 $(eval $(call BuildPackage,libopenssl-conf))
 $(eval $(call BuildPackage,libopenssl-afalg))
@@ -435,3 +453,4 @@ $(eval $(call BuildPackage,libopenssl-de
 $(eval $(call BuildPackage,libopenssl-legacy))
 $(eval $(call BuildPackage,libopenssl-padlock))
 $(eval $(call BuildPackage,openssl-util))
+$(eval $(call HostBuild))
