From 67e1748c9006dfa3e0126f4877381f926fbcfe14 Mon Sep 17 00:00:00 2001
From: Shayne Chen <shayne.chen@mediatek.com>
Date: Thu, 19 Oct 2023 10:48:11 +0800
Subject: [PATCH 054/206] mtk: wpa_supplicant: correctly get assoc frequency

Signed-off-by: Shayne Chen <shayne.chen@mediatek.com>
---
 src/drivers/driver_nl80211_event.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/drivers/driver_nl80211_event.c b/src/drivers/driver_nl80211_event.c
index 3fdaa54b9..d6d69aa7c 100644
--- a/src/drivers/driver_nl80211_event.c
+++ b/src/drivers/driver_nl80211_event.c
@@ -331,6 +331,7 @@ static void mlme_event_assoc(struct wpa_driver_nl80211_data *drv,
 			   wpa_ssid_txt(drv->ssid, drv->ssid_len));
 	}
 
+	drv->assoc_freq = nl80211_get_assoc_freq(drv);
 	event.assoc_info.freq = drv->assoc_freq;
 	drv->first_bss->flink->freq = drv->assoc_freq;
 
-- 
2.45.2

