From 14f48d28709d46425bdec7f02b71a6256c795035 Mon Sep 17 00:00:00 2001
From: Rex Lu <rex.lu@mediatek.com>
Date: Thu, 16 Oct 2025 15:33:05 +0800
Subject: [PATCH] backport-6-17-refactor-mtk-wed-device

---
 include/linux/soc/mediatek/mtk_wed.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/include/linux/soc/mediatek/mtk_wed.h b/include/linux/soc/mediatek/mtk_wed.h
index e153e76..55b7c6d 100644
--- a/include/linux/soc/mediatek/mtk_wed.h
+++ b/include/linux/soc/mediatek/mtk_wed.h
@@ -170,6 +170,7 @@ struct mtk_wed_device {
 		bool wcid_512;
 		enum mtk_wed_hwrro_mode hw_rro;
 		bool msi;
+		bool hif2;
 
 		u16 token_start;
 		unsigned int nbuf;
-- 
2.45.2

