diff --git a/arch/arm64/boot/dts/mediatek/mt7988a-rfb.dts b/arch/arm64/boot/dts/mediatek/mt7988a-rfb.dts
index f55a91f..be332d3 100644
--- a/arch/arm64/boot/dts/mediatek/mt7988a-rfb.dts
+++ b/arch/arm64/boot/dts/mediatek/mt7988a-rfb.dts
@@ -199,6 +199,8 @@ &i2c2 {
 };
 
 &pcie0 {
+	wifi-reset-gpios = <&pio 7 GPIO_ACTIVE_HIGH>;
+	wifi-reset-msleep = <100>;
 	status = "okay";
 };
 
diff --git a/arch/arm64/boot/dts/mediatek/mt7988a.dtsi b/arch/arm64/boot/dts/mediatek/mt7988a.dtsi
index cb342eb..c7f58d2 100644
--- a/arch/arm64/boot/dts/mediatek/mt7988a.dtsi
+++ b/arch/arm64/boot/dts/mediatek/mt7988a.dtsi
@@ -583,8 +583,7 @@ mux {
 			pcie0_pins: pcie0-pins {
 				mux {
 					function = "pcie";
-					groups = "pcie_2l_0_pereset", "pcie_clk_req_n0_0",
-						 "pcie_wake_n0_0";
+					groups = "pcie_2l_0_pereset", "pcie_clk_req_n0_0";
 				};
 			};
 
