Subject: [PATCH] net-phy-add-aeonsemi-as21xxx-support

diff --git a/drivers/net/phy/Kconfig b/drivers/net/phy/Kconfig
index 223db6f..83757fc 100644
--- a/drivers/net/phy/Kconfig
+++ b/drivers/net/phy/Kconfig
@@ -146,6 +146,8 @@ source "drivers/net/phy/airoha/an8855/Kconfig"
 
 comment "MII PHY device drivers"
 
+source "drivers/net/phy/as21xxx/Kconfig"
+
 config AIROHA_AN8801_PHY
 	tristate "Drivers for Airoha AN8801 Gigabit PHYs"
 	help
diff --git a/drivers/net/phy/Makefile b/drivers/net/phy/Makefile
index 603b37d..b917b57 100644
--- a/drivers/net/phy/Makefile
+++ b/drivers/net/phy/Makefile
@@ -55,6 +55,7 @@ obj-$(CONFIG_AIR_AN8855_PHY)		+= air_an8855.o
 obj-$(CONFIG_AIR_EN8811H_PHY)   += air_en8811h.o
 obj-$(CONFIG_AMD_PHY)		+= amd.o
 obj-$(CONFIG_AQUANTIA_PHY)	+= aquantia/
+obj-$(CONFIG_AS21XXX_PHY)  += as21xxx/
 obj-$(CONFIG_AX88796B_PHY)	+= ax88796b.o
 obj-$(CONFIG_BCM54140_PHY)	+= bcm54140.o
 obj-$(CONFIG_BCM63XX_PHY)	+= bcm63xx.o
-- 
2.45.2

