From 614490716808ca0b3a9754f5c5370384da461ffd Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Wed, 22 Oct 2025 15:20:04 +0800
Subject: [PATCH 239/254] mtk: hostapd: align QCA's prpl link add command
 format

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 hostapd/hostapd_cli.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/hostapd/hostapd_cli.c b/hostapd/hostapd_cli.c
index 9d8bc823b..a2994eca2 100644
--- a/hostapd/hostapd_cli.c
+++ b/hostapd/hostapd_cli.c
@@ -2215,6 +2215,8 @@ static const struct hostapd_cli_cmd hostapd_cli_commands[] = {
 		" = <ac> [cwmin=] [cwmax=] [aifs=] [txop_limit=]"},
 	{ "link_add", hostapd_cli_cmd_link_add, NULL,
 		" = Add a new link to a MLD AP"},
+	{ "mld_link_add", hostapd_cli_cmd_link_add, NULL,
+		" = Add a new link to a MLD AP (QCA ver.)"},
 	{ "link_remove", hostapd_cli_cmd_link_remove, NULL,
 		" [count=<count>] = Remove affiliated link of a MLD AP"},
 	{ "set_attlm", hostapd_cli_cmd_set_attlm, NULL,
-- 
2.45.2

