--- a/arch/arm64/boot/dts/mediatek/mt7986a-rfb.dts
+++ b/arch/arm64/boot/dts/mediatek/mt7986a-rfb.dts
@@ -28,6 +28,18 @@
 		reg = <0 0x40000000 0 0x40000000>;
 	};
 
+ 	pwm_led: pwm-led {
+		compatible = "pwm-leds";
+		status = "disabled";
+
+		led {
+			pwms = <&pwm 0 50000 0>;
+			max-brightness = <255>;
+			active-low;
+			linux,default-trigger = "default-on";
+		};
+	};
+
 	reg_1p8v: regulator-1p8v {
 		compatible = "regulator-fixed";
 		regulator-name = "fixed-1.8V";
@@ -291,8 +303,17 @@
 			groups = "i2c";
 		};
 	};
+
+ 	pwm_pins: pwm-pins {
+		mux {
+			function = "pwm";
+			groups = "pwm0";
+		};
+	};
 };
 
+
+
 &spi0 {
 	pinctrl-names = "default";
 	pinctrl-0 = <&spi_flash_pins>;
@@ -363,6 +384,12 @@
 	};
 };
 
+&pwm_led {
+ 	pinctrl-names = "default";
+	pinctrl-0 = <&pwm_pins>;
+	status = "okay";
+};
+
 &uart0 {
 	status = "okay";
 };
