--- a/drivers/net/phy/mediatek/mtk-2p5ge.c
+++ b/drivers/net/phy/mediatek/mtk-2p5ge.c
@@ -218,24 +218,24 @@ static inline void reg_writew(struct mtk
 }
 
 static inline void reg_modify(struct mtk_i2p5ge_phy_priv *priv, u32 offset,
-			      u32 mask, u32 val)
+			      u32 mask, u32 set)
 {
 	u32 reg = reg_readl(priv, offset);
 
-	reg = (reg & ~mask) | (val & mask);
+	reg = (reg & ~mask) | set;
 	reg_writel(priv, offset, reg);
 }
 
 static inline void reg_set_bits(struct mtk_i2p5ge_phy_priv *priv, u32 offset,
-				u32 bits)
+				u32 val)
 {
-	reg_modify(priv, offset, bits, bits);
+	reg_modify(priv, offset, 0, val);
 }
 
 static inline void reg_clear_bits(struct mtk_i2p5ge_phy_priv *priv, u32 offset,
-				  u32 bits)
+				  u32 val)
 {
-	reg_modify(priv, offset, bits, 0);
+	reg_modify(priv, offset, val, 0);
 }
 
 static ssize_t mtk_2p5ge_attr_show(struct device *dev,
@@ -411,10 +411,12 @@ static int mt7987_2p5ge_phy_load_fw(stru
 		goto release_fw;
 	}
 
+	phy_lock_mdio_bus(phydev);
+
 	/* Force 2.5Gphy back to AN state */
-	phy_set_bits(phydev, MII_BMCR, BMCR_RESET);
+	__phy_set_bits(phydev, MII_BMCR, BMCR_RESET);
 	usleep_range(5000, 6000);
-	phy_set_bits(phydev, MII_BMCR, BMCR_PDOWN);
+	__phy_set_bits(phydev, MII_BMCR, BMCR_PDOWN);
 
 	reg = readw(apb_base + SW_RESET);
 	writew(reg & ~MD32_RESTART_EN_CLEAR, apb_base + SW_RESET);
@@ -455,17 +457,17 @@ static int mt7987_2p5ge_phy_load_fw(stru
 		dev_err(dev, "Fail to enable 2.5Gbps module clock: %d\n", ret);
 
 	/* Disable AN */
-	phy_clear_bits(phydev, MII_BMCR, BMCR_ANENABLE);
+	__phy_clear_bits(phydev, MII_BMCR, BMCR_ANENABLE);
 
 	/* Force to run at 2.5G speed */
-	phy_modify_mmd(phydev, MDIO_MMD_VEND1, MTK_PHY_AN_FORCE_SPEED_REG,
-		       MTK_PHY_MASTER_FORCE_SPEED_SEL_MASK,
-		       MTK_PHY_MASTER_FORCE_SPEED_SEL_EN |
-		       FIELD_PREP(MTK_PHY_MASTER_FORCE_SPEED_SEL_MASK, 0x1b));
-
-	phy_set_bits_mmd(phydev, MDIO_MMD_VEND1, MTK_PHY_LINK_STATUS_RELATED,
-			 MTK_PHY_BYPASS_LINK_STATUS_OK |
-			 MTK_PHY_FORCE_LINK_STATUS_HCD);
+	__phy_modify_mmd(phydev, MDIO_MMD_VEND1, MTK_PHY_AN_FORCE_SPEED_REG,
+			 MTK_PHY_MASTER_FORCE_SPEED_SEL_MASK,
+			 MTK_PHY_MASTER_FORCE_SPEED_SEL_EN |
+			 FIELD_PREP(MTK_PHY_MASTER_FORCE_SPEED_SEL_MASK, 0x1b));
+
+	__phy_set_bits_mmd(phydev, MDIO_MMD_VEND1, MTK_PHY_LINK_STATUS_RELATED,
+			   MTK_PHY_BYPASS_LINK_STATUS_OK |
+			   MTK_PHY_FORCE_LINK_STATUS_HCD);
 
 	/* Set xbz, pma and rx as "do not reset" in order to input DSP code. */
 	reg_set_bits(priv, MTK_2P5GPHY_PMD_REG + DO_NOT_RESET,
@@ -533,10 +535,12 @@ static int mt7987_2p5ge_phy_load_fw(stru
 
 	reg = reg_readw(priv, MTK_2P5GPHY_MCU_CSR + MD32_EN_CFG);
 	reg_writew(priv, MTK_2P5GPHY_MCU_CSR + MD32_EN_CFG, reg | MD32_EN);
-	phy_set_bits(phydev, MII_BMCR, BMCR_RESET);
+	__phy_set_bits(phydev, MII_BMCR, BMCR_RESET);
 	/* We need a delay here to stabilize initialization of MCU */
 	usleep_range(7000, 8000);
 
+	phy_unlock_mdio_bus(phydev);
+
 release_fw:
 	release_firmware(fw);
 
@@ -565,16 +569,18 @@ static int mt7988_2p5ge_phy_load_fw(stru
 		goto release_fw;
 	}
 
+	phy_lock_mdio_bus(phydev);
+
 	reg = reg_readw(priv, MTK_2P5GPHY_MCU_CSR + MD32_EN_CFG);
 	if (reg & MD32_EN) {
-		phy_set_bits(phydev, MII_BMCR, BMCR_RESET);
+		__phy_set_bits(phydev, MII_BMCR, BMCR_RESET);
 		usleep_range(10000, 11000);
 	}
-	phy_set_bits(phydev, MII_BMCR, BMCR_PDOWN);
+	__phy_set_bits(phydev, MII_BMCR, BMCR_PDOWN);
 
 	/* Write magic number to safely stall MCU */
-	phy_write_mmd(phydev, MDIO_MMD_VEND1, MTK_PHY_HOST_CMD1, 0x1100);
-	phy_write_mmd(phydev, MDIO_MMD_VEND1, MTK_PHY_HOST_CMD2, 0x00df);
+	__phy_write_mmd(phydev, MDIO_MMD_VEND1, MTK_PHY_HOST_CMD1, 0x1100);
+	__phy_write_mmd(phydev, MDIO_MMD_VEND1, MTK_PHY_HOST_CMD2, 0x00df);
 
 	for (i = 0; i < MT7988_2P5GE_PMB_FW_SIZE - 1; i += 4)
 		reg_writel(priv, MTK_2P5GPHY_PMB_FW + i,
@@ -582,10 +588,12 @@ static int mt7988_2p5ge_phy_load_fw(stru
 
 	reg_writew(priv, MTK_2P5GPHY_MCU_CSR + MD32_EN_CFG, reg & ~MD32_EN);
 	reg_writew(priv, MTK_2P5GPHY_MCU_CSR + MD32_EN_CFG, reg | MD32_EN);
-	phy_set_bits(phydev, MII_BMCR, BMCR_RESET);
+	__phy_set_bits(phydev, MII_BMCR, BMCR_RESET);
 	/* We need a delay here to stabilize initialization of MCU */
 	usleep_range(7000, 8000);
 
+	phy_unlock_mdio_bus(phydev);
+
 	mtk_2p5ge_fw_info(dev, priv, fw, MT7988_2P5GE_PMB_FW_SIZE);
 
 release_fw:
