From c3f25accc62f18adb22abc256add6c954fa415a0 Mon Sep 17 00:00:00 2001
From: Tim-cy Yang <Tim-cy.Yang@mediatek.com>
Date: Fri, 20 Mar 2026 16:16:49 +0800
Subject: [PATCH] soc: mediatek: mtk-devapc: add mt798x support

---
 drivers/soc/mediatek/mtk-devapc.c | 150 +++++++++++++++++++++++++++---
 1 file changed, 139 insertions(+), 11 deletions(-)

diff --git a/drivers/soc/mediatek/mtk-devapc.c b/drivers/soc/mediatek/mtk-devapc.c
index f54c966138b5..7256834e37a5 100644
--- a/drivers/soc/mediatek/mtk-devapc.c
+++ b/drivers/soc/mediatek/mtk-devapc.c
@@ -3,6 +3,7 @@
  * Copyright (C) 2020 MediaTek Inc.
  */
 
+#include <linux/arm-smccc.h>
 #include <linux/clk.h>
 #include <linux/interrupt.h>
 #include <linux/iopoll.h>
@@ -11,11 +12,15 @@
 #include <linux/of.h>
 #include <linux/of_irq.h>
 #include <linux/of_address.h>
+#include <linux/soc/mediatek/mtk_sip_svc.h>
 
 #define VIO_MOD_TO_REG_IND(m)	((m) / 32)
 #define VIO_MOD_TO_REG_OFF(m)	((m) % 32)
 
-struct mtk_devapc_vio_dbgs {
+#define MTK_SIP_DAPC_GET_SEC_CFG	MTK_SIP_SMC_CMD(0x5A0)
+#define NO_PROTECTION			0
+
+struct mtk_devapc_vio_dbgs_v1 {
 	union {
 		u32 vio_dbg0;
 		struct {
@@ -31,12 +36,29 @@ struct mtk_devapc_vio_dbgs {
 	u32 vio_dbg1;
 };
 
+struct mtk_devapc_vio_dbgs_v2 {
+	union {
+		u32 vio_dbg0;
+		struct {
+			u32 dmnid:6;
+			u32 vio_w:1;
+			u32 vio_r:1;
+		} dbg0_bits;
+	};
+
+	u32 vio_dbg1;
+	u32 vio_dbg2;
+	u32 vio_dbg3;
+};
+
 struct mtk_devapc_regs_ofs {
 	/* reg offset */
 	u32 vio_mask_offset;
 	u32 vio_sta_offset;
 	u32 vio_dbg0_offset;
 	u32 vio_dbg1_offset;
+	u32 vio_dbg2_offset;
+	u32 vio_dbg3_offset;
 	u32 apc_con_offset;
 	u32 vio_shift_sta_offset;
 	u32 vio_shift_sel_offset;
@@ -46,6 +68,8 @@ struct mtk_devapc_regs_ofs {
 struct mtk_devapc_data {
 	/* numbers of violation index */
 	u32 vio_idx_num;
+	u32 version;
+	u32 nsec_access;
 	const struct mtk_devapc_regs_ofs *regs_ofs;
 };
 
@@ -154,13 +178,9 @@ static int devapc_sync_vio_dbg(struct mtk_devapc_context *ctx)
 	return true;
 }
 
-/*
- * devapc_extract_vio_dbg - extract full violation information after doing
- *                          shift mechanism.
- */
-static void devapc_extract_vio_dbg(struct mtk_devapc_context *ctx)
+static void devapc_extract_vio_dbg_v1(struct mtk_devapc_context *ctx)
 {
-	struct mtk_devapc_vio_dbgs vio_dbgs;
+	struct mtk_devapc_vio_dbgs_v1 vio_dbgs;
 	void __iomem *vio_dbg0_reg;
 	void __iomem *vio_dbg1_reg;
 
@@ -181,6 +201,46 @@ static void devapc_extract_vio_dbg(struct mtk_devapc_context *ctx)
 		 vio_dbgs.vio_dbg1);
 }
 
+static void devapc_extract_vio_dbg_v2(struct mtk_devapc_context *ctx)
+{
+	struct mtk_devapc_vio_dbgs_v2 vio_dbgs;
+	void __iomem *vio_dbg0_reg;
+	void __iomem *vio_dbg1_reg;
+	void __iomem *vio_dbg2_reg;
+	void __iomem *vio_dbg3_reg;
+
+	vio_dbg0_reg = ctx->infra_base + ctx->data->regs_ofs->vio_dbg0_offset;
+	vio_dbg1_reg = ctx->infra_base + ctx->data->regs_ofs->vio_dbg1_offset;
+	vio_dbg2_reg = ctx->infra_base + ctx->data->regs_ofs->vio_dbg2_offset;
+	vio_dbg3_reg = ctx->infra_base + ctx->data->regs_ofs->vio_dbg3_offset;
+
+	vio_dbgs.vio_dbg0 = readl(vio_dbg0_reg);
+	vio_dbgs.vio_dbg1 = readl(vio_dbg1_reg);
+	vio_dbgs.vio_dbg2 = readl(vio_dbg2_reg);
+	vio_dbgs.vio_dbg3 = readl(vio_dbg3_reg);
+
+	if (vio_dbgs.dbg0_bits.vio_w)
+		dev_info(ctx->dev, "Write Violation\n");
+	else if (vio_dbgs.dbg0_bits.vio_r)
+		dev_info(ctx->dev, "Read Violation\n");
+
+	dev_info(ctx->dev, "Bus ID:0x%x, Dom ID:0x%x, Vio Addr:0x%llx\n",
+		 vio_dbgs.vio_dbg1, vio_dbgs.dbg0_bits.dmnid,
+		 ((u64)vio_dbgs.vio_dbg3 << 32) | vio_dbgs.vio_dbg2);
+}
+
+/*
+ * devapc_extract_vio_dbg - extract full violation information after doing
+ *                          shift mechanism.
+ */
+static void devapc_extract_vio_dbg(struct mtk_devapc_context *ctx)
+{
+	if (ctx->data->version == 1)
+		devapc_extract_vio_dbg_v1(ctx);
+	else
+		devapc_extract_vio_dbg_v2(ctx);
+}
+
 /*
  * devapc_violation_irq - the devapc Interrupt Service Routine (ISR) will dump
  *                        violation information including which master violates
@@ -229,16 +289,58 @@ static const struct mtk_devapc_regs_ofs devapc_regs_ofs_mt6779 = {
 	.vio_shift_con_offset = 0xF20,
 };
 
+static const struct mtk_devapc_regs_ofs devapc_regs_ofs_mt7986 = {
+	.vio_mask_offset = 0x0,
+	.vio_sta_offset = 0x400,
+	.vio_dbg0_offset = 0x900,
+	.vio_dbg1_offset = 0x904,
+	.vio_dbg2_offset = 0x908,
+	.vio_dbg3_offset = 0x90C,
+	.apc_con_offset = 0xF00,
+	.vio_shift_sta_offset = 0xF20,
+	.vio_shift_sel_offset = 0xF30,
+	.vio_shift_con_offset = 0xF10,
+};
+
 static const struct mtk_devapc_data devapc_mt6779 = {
 	.vio_idx_num = 511,
+	.version = 1,
+	.nsec_access = 1,
 	.regs_ofs = &devapc_regs_ofs_mt6779,
 };
 
 static const struct mtk_devapc_data devapc_mt8186 = {
 	.vio_idx_num = 519,
+	.version = 1,
+	.nsec_access = 1,
 	.regs_ofs = &devapc_regs_ofs_mt6779,
 };
 
+static const struct mtk_devapc_data devapc_mt7981 = {
+	.vio_idx_num = 293,
+	.version = 2,
+	.regs_ofs = &devapc_regs_ofs_mt7986,
+};
+
+static const struct mtk_devapc_data devapc_mt7986 = {
+	.vio_idx_num = 270,
+	.version = 2,
+	.regs_ofs = &devapc_regs_ofs_mt7986,
+};
+
+static const struct mtk_devapc_data devapc_mt7987 = {
+	.vio_idx_num = 283,
+	.version = 2,
+	.nsec_access = 1,
+	.regs_ofs = &devapc_regs_ofs_mt7986,
+};
+
+static const struct mtk_devapc_data devapc_mt7988 = {
+	.vio_idx_num = 298,
+	.version = 2,
+	.regs_ofs = &devapc_regs_ofs_mt7986,
+};
+
 static const struct of_device_id mtk_devapc_dt_match[] = {
 	{
 		.compatible = "mediatek,mt6779-devapc",
@@ -246,6 +348,18 @@ static const struct of_device_id mtk_devapc_dt_match[] = {
 	}, {
 		.compatible = "mediatek,mt8186-devapc",
 		.data = &devapc_mt8186,
+	}, {
+		.compatible = "mediatek,mt7981-devapc",
+		.data = &devapc_mt7981,
+	}, {
+		.compatible = "mediatek,mt7986-devapc",
+		.data = &devapc_mt7986,
+	}, {
+		.compatible = "mediatek,mt7987-devapc",
+		.data = &devapc_mt7987,
+	}, {
+		.compatible = "mediatek,mt7988-devapc",
+		.data = &devapc_mt7988,
 	}, {
 	},
 };
@@ -255,6 +369,7 @@ static int mtk_devapc_probe(struct platform_device *pdev)
 {
 	struct device_node *node = pdev->dev.of_node;
 	struct mtk_devapc_context *ctx;
+	struct arm_smccc_res res;
 	u32 devapc_irq;
 	int ret;
 
@@ -268,6 +383,17 @@ static int mtk_devapc_probe(struct platform_device *pdev)
 	ctx->data = of_device_get_match_data(&pdev->dev);
 	ctx->dev = &pdev->dev;
 
+	/* check if violation can be handled by kernel */
+	if (!ctx->data->nsec_access) {
+		arm_smccc_smc(MTK_SIP_DAPC_GET_SEC_CFG, 0, 0, 0, 0, 0, 0, 0,
+			      &res);
+		if (res.a0 != NO_PROTECTION) {
+			dev_info(&pdev->dev,
+				 "violation cannot be handled in normal world\n");
+			return 0;
+		}
+	}
+
 	ctx->infra_base = of_iomap(node, 0);
 	if (!ctx->infra_base)
 		return -EINVAL;
@@ -278,10 +404,12 @@ static int mtk_devapc_probe(struct platform_device *pdev)
 		goto err;
 	}
 
-	ctx->infra_clk = devm_clk_get_enabled(&pdev->dev, "devapc-infra-clock");
-	if (IS_ERR(ctx->infra_clk)) {
-		ret = -EINVAL;
-		goto err;
+	if (ctx->data->version == 1) {
+		ctx->infra_clk = devm_clk_get_enabled(&pdev->dev, "devapc-infra-clock");
+		if (IS_ERR(ctx->infra_clk)) {
+			ret = -EINVAL;
+			goto err;
+		}
 	}
 
 	ret = devm_request_irq(&pdev->dev, devapc_irq, devapc_violation_irq,
-- 
2.45.2

