--- /dev/null
+++ b/arch/arm64/boot/dts/mediatek/mt7988d-rfb.dts
@@ -0,0 +1,82 @@
+// SPDX-License-Identifier: (GPL-2.0 OR MIT)
+/*
+ * Copyright (C) 2024 MediaTek Inc.
+ * Author: Sam.Shih <sam.shih@mediatek.com>
+ */
+
+/dts-v1/;
+#include "mt7988a-rfb.dts"
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+/ {
+	model = "MediaTek MT7988A Reference Board";
+	compatible = "mediatek,mt7988d-rfb",
+		     "mediatek,mt7988d",
+		     "mediatek,mt7988";
+
+	chosen {
+		bootargs = "console=ttyS0,115200n1 loglevel=6  \
+			    earlycon=uart8250,mmio32,0x11000000 \
+			    pci=pcie_bus_perf ubi.block=0,firmware root=/dev/fit0 \
+			    rootwait";
+	};
+
+	cpus {
+		/delete-node/ cpu@3;
+	};
+
+	memory {
+		reg = <0 0x40000000 0 0x40000000>;
+	};
+
+	gpio-keys {
+		compatible = "gpio-keys";
+
+		reset {
+			label = "reset";
+			linux,code = <KEY_RESTART>;
+			gpios = <&pio 13 GPIO_ACTIVE_LOW>;
+		};
+
+		wps {
+			label = "wps";
+			linux,code = <KEY_WPS_BUTTON>;
+			gpios = <&pio 14 GPIO_ACTIVE_LOW>;
+		};
+	};
+};
+
+/*
+ * MT7988A have 4 pcies:
+ * -pcie0: pcie@11300000 - Gen3 2 lanes
+ * -pcie1: pcie@11310000 - Gen3 2 lanes
+ * -pcie2: pcie@11280000 - Gen3 1 lane (shared with u3 port of ssusb0)
+ * -pcie3: pcie@11290000 - Gen3 1 lane
+ * and 2 usb controllers
+ * -ssusb0: usb@11190000 (u2 port and u3 port, u3 port shared with pcie2)
+ * -ssusb1: usb@11200000 (u2 port and u3 port)
+ *
+ * MT7988D only support 2 pcies:
+ * -pcie0: pcie@11300000 - Gen3 2 lanes
+ * -pcie3: pcie@11290000 - Gen3 1 lane
+ * and 1 usb controller
+ * -ssusb1: usb@11200000 (u2 port and u3 port)
+ */
+
+&pcie1 {
+	status = "disabled";
+};
+
+&pcie2 {
+	status = "disabled";
+};
+
+&ssusb0 {
+	status = "disabled";
+};
+
+&xsphy {
+	status = "disabled";
+};
+
--- /dev/null
+++ b/arch/arm64/boot/dts/mediatek/mt7988d-rfb-eth0-gsw.dtso
@@ -0,0 +1,177 @@
+// SPDX-License-Identifier: (GPL-2.0 OR MIT)
+/*
+ * Copyright (C) 2025 MediaTek Inc.
+ * Author: Neal.Yen <neal.yen@mediatek.com>
+ */
+
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/interrupt-controller/arm-gic.h>
+#include <dt-bindings/interrupt-controller/irq.h>
+#include <dt-bindings/leds/common.h>
+
+/ {
+
+	fragment@0 {
+		target-path = "/";
+		__overlay__ {
+			compatible = "mediatek,mt7988d-rfb-gsw",
+						 "mediatek,mt7988d",
+						 "mediatek,mt7988";
+		};
+	};
+
+	fragment@1 {
+		target = <&switch>;
+		__overlay__ {
+			status = "disabled";
+		};
+	};
+
+	fragment@2 {
+		target = <&ethwarp>;
+		__overlay__ {
+			compatible = "mediatek,mt7988-ethwarp", "syscon";
+		};
+	};
+
+	fragment@3 {
+		target = <&gsw>;
+		__overlay__ {
+			compatible = "mediatek,mt753x", "mediatek,mt7988-gsw-switch", "syscon";
+			mediatek,sysctrl = <&ethwarp>;
+			mediatek,mdio = <&mdio_bus>;
+			mediatek,portmap = "llllw";
+			mediatek,mdio_master_pinmux = <1>;
+			interrupt-parent = <&gic>;
+			interrupts = <GIC_SPI 209 IRQ_TYPE_LEVEL_HIGH>;
+			status = "okay";
+
+			port6: port@6 {
+				compatible = "mediatek,mt753x-port";
+				mediatek,ssc-on;
+				phy-mode = "10gbase-kr";
+				reg = <6>;
+				fixed-link {
+					speed = <10000>;
+					full-duplex;
+				};
+			};
+
+			mdio1: mdio-bus {
+				#address-cells = <1>;
+				#size-cells = <0>;
+				mediatek,pio = <&pio>;
+
+				gsw_phy0: ethernet-phy@0 {
+					compatible = "ethernet-phy-ieee802.3-c22";
+					reg = <0>;
+					phy-mode = "internal";
+					pinctrl-names = "gbe-led";
+					pinctrl-0 = <&gbe0_led0_pins>;
+					nvmem-cells = <&phy_calibration_p0>;
+					nvmem-cell-names = "phy-cal-data";
+
+					leds {
+						#address-cells = <1>;
+						#size-cells = <0>;
+
+						gsw_phy0_led0: gsw-phy0-led0@0 {
+							reg = <0>;
+							function = LED_FUNCTION_LAN;
+							status = "disabled";
+						};
+
+						gsw_phy0_led1: gsw-phy0-led1@1 {
+							reg = <1>;
+							function = LED_FUNCTION_LAN;
+							status = "disabled";
+						};
+					};
+				};
+
+				gsw_phy1: ethernet-phy@1 {
+					compatible = "ethernet-phy-ieee802.3-c22";
+					reg = <1>;
+					phy-mode = "internal";
+					pinctrl-names = "gbe-led";
+					pinctrl-0 = <&gbe1_led0_pins>;
+					nvmem-cells = <&phy_calibration_p1>;
+					nvmem-cell-names = "phy-cal-data";
+
+					leds {
+						#address-cells = <1>;
+						#size-cells = <0>;
+
+						gsw_phy1_led0: gsw-phy1-led0@0 {
+							reg = <0>;
+							function = LED_FUNCTION_LAN;
+							status = "disabled";
+						};
+
+						gsw_phy1_led1: gsw-phy1-led1@1 {
+							reg = <1>;
+							function = LED_FUNCTION_LAN;
+							status = "disabled";
+						};
+					};
+				};
+
+				gsw_phy2: ethernet-phy@2 {
+					compatible = "ethernet-phy-ieee802.3-c22";
+					reg = <2>;
+					phy-mode = "internal";
+					pinctrl-names = "gbe-led";
+					pinctrl-0 = <&gbe2_led0_pins>;
+					nvmem-cells = <&phy_calibration_p2>;
+					nvmem-cell-names = "phy-cal-data";
+
+					leds {
+						#address-cells = <1>;
+						#size-cells = <0>;
+
+						gsw_phy2_led0: gsw-phy2-led0@0 {
+							reg = <0>;
+							function = LED_FUNCTION_LAN;
+							status = "disabled";
+						};
+
+						gsw_phy2_led1: gsw-phy2-led1@1 {
+							reg = <1>;
+							function = LED_FUNCTION_LAN;
+							status = "disabled";
+						};
+					};
+				};
+
+				gsw_phy3: ethernet-phy@3 {
+					compatible = "ethernet-phy-ieee802.3-c22";
+					reg = <3>;
+					phy-mode = "internal";
+					pinctrl-names = "gbe-led";
+					pinctrl-0 = <&gbe3_led0_pins>;
+					nvmem-cells = <&phy_calibration_p3>;
+					nvmem-cell-names = "phy-cal-data";
+
+					leds {
+						#address-cells = <1>;
+						#size-cells = <0>;
+
+						gsw_phy3_led0: gsw-phy3-led0@0 {
+							reg = <0>;
+							function = LED_FUNCTION_LAN;
+							status = "disabled";
+						};
+
+						gsw_phy3_led1: gsw-phy3-led1@1 {
+							reg = <1>;
+							function = LED_FUNCTION_LAN;
+							status = "disabled";
+						};
+					};
+				};
+			};
+		};
+	};
+};
\ No newline at end of file
--- /dev/null
+++ b/arch/arm64/boot/dts/mediatek/mt7988d-rfb-eth2-an8831x.dtso
@@ -0,0 +1,70 @@
+// SPDX-License-Identifier: (GPL-2.0 OR MIT)
+/*
+ * Copyright (C) 2022 MediaTek Inc.
+ * Author: Sam.Shih <sam.shih@mediatek.com>
+ */
+
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/gpio.h>
+
+/ {
+	compatible = "mediatek,mt7988a-rfb", "mediatek,mt7988d";
+
+	fragment@0 {
+		target = <&mdio_bus>;
+		__overlay__ {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			clock-frequency = <12500000>;
+			reset-gpios = <&pio 3 GPIO_ACTIVE_LOW>;
+			reset-delay-us = <200000>;
+			reset-post-delay-us = <350000>;
+
+			/* external Airoha AN8831X */
+			phy24: ethernet-phy@24 {
+				reg = <24>;
+				compatible = "ethernet-phy-ieee802.3-c45";
+				firmware-name = "as21x1x_fw.bin";
+
+				leds {
+					#address-cells = <1>;
+					#size-cells = <0>;
+
+					phy24_led0: an8831x_phy24_led0@0 {
+						reg = <0>;
+						linux,default-trigger = "netdev";
+						active-high;
+						status = "okay";
+					};
+
+					phy24_led1: an8831x_phy24_led1@1 {
+						reg = <1>;
+						linux,default-trigger = "netdev";
+						active-high;
+						status = "okay";
+					};
+				};
+			};
+		};
+	};
+
+	fragment@1 {
+		target = <&gmac2>;
+		__overlay__ {
+			phy-mode = "usxgmii";
+			phy-connection-type = "usxgmii";
+			phy = <&phy24>;
+			status = "okay";
+		};
+	};
+
+	fragment@2 {
+		target = <&eth>;
+		__overlay__ {
+			pinctrl-0 = <&mdio0_pins>;
+			pinctrl-names = "default";
+		};
+	};
+};
--- /dev/null
+++ b/arch/arm64/boot/dts/mediatek/mt7988d-rfb-eth2-sfp.dtso
@@ -0,0 +1,47 @@
+// SPDX-License-Identifier: (GPL-2.0 OR MIT)
+/*
+ * Copyright (C) 2022 MediaTek Inc.
+ * Author: Sam.Shih <sam.shih@mediatek.com>
+ */
+
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/gpio.h>
+
+/ {
+	compatible = "mediatek,mt7988d-rfb", "mediatek,mt7988d";
+
+	fragment@0 {
+		target = <&i2c1>;
+		__overlay__ {
+			pinctrl-names = "default";
+			pinctrl-0 = <&i2c1_sfp_pins>;
+			status = "okay";
+		};
+	};
+
+	fragment@1 {
+		target-path = "/";
+		__overlay__ {
+			sfp_esp0: sfp@0 {
+				compatible = "sff,sfp";
+				i2c-bus = <&i2c1>;
+				mod-def0-gpios = <&pio 0 GPIO_ACTIVE_LOW>;
+				los-gpios = <&pio 30 GPIO_ACTIVE_HIGH>;
+				tx-disable-gpios = <&pio 29 GPIO_ACTIVE_HIGH>;
+				maximum-power-milliwatt = <3000>;
+			};
+		};
+	};
+
+	fragment@2 {
+		target = <&gmac2>;
+		__overlay__ {
+			phy-mode = "10gbase-r";
+			managed = "in-band-status";
+			sfp = <&sfp_esp0>;
+			status = "okay";
+		};
+	};
+};
