From 872b5bd1b3975ade934902f03f9ea58855b62092 Mon Sep 17 00:00:00 2001
From: "neal.yen" <neal.yen@mediatek.com>
Date: Tue, 3 Mar 2026 14:22:14 +0800
Subject: [PATCH] net: dsa: fix VLAN deletion to override the FID

---
 drivers/net/dsa/mt7530.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/net/dsa/mt7530.c b/drivers/net/dsa/mt7530.c
index f8e6ca3..74bcefb 100644
--- a/drivers/net/dsa/mt7530.c
+++ b/drivers/net/dsa/mt7530.c
@@ -1747,7 +1747,7 @@ mt7530_hw_vlan_del(struct mt7530_priv *priv,
 
 	if (new_members) {
 		val = IVL_MAC | VTAG_EN | PORT_MEM(new_members) |
-		      VLAN_VALID;
+		      FID(FID_BRIDGED) | VLAN_VALID;
 		mt7530_write(priv, MT7530_VAWD1, val);
 	} else {
 		mt7530_write(priv, MT7530_VAWD1, 0);
-- 
2.45.2

