--- a/tools/Makefile
+++ b/tools/Makefile
@@ -146,9 +148,9 @@ else
 endif
 
 ifneq ($(CONFIG_CCACHE)$(CONFIG_SDK),)
-$(foreach tool, $(filter-out zstd zlib xz pkgconf patch ninja meson libressl expat cmake,$(tools-y)), $(eval $(curdir)/$(tool)/compile += $(curdir)/ccache/compile))
-tools-y += ccache
-$(curdir)/ccache/compile := $(curdir)/cmake/compile
+$(foreach tool, $(filter-out zstd zlib xz pkgconf patch ninja meson libressl expat cmake xxhash,$(tools-y)), $(eval $(curdir)/$(tool)/compile += $(curdir)/ccache/compile))
+tools-y += ccache xxhash
+$(curdir)/ccache/compile := $(curdir)/cmake/compile $(curdir)/xxhash/compile
 endif
 
 # in case there is no patch tool on the host we need to make patch tool a
