--- a/target/linux/mediatek/files-6.6/arch/arm64/boot/dts/mediatek/mt7988a.dtsi
+++ b/target/linux/mediatek/files-6.6/arch/arm64/boot/dts/mediatek/mt7988a.dtsi
@@ -530,6 +530,14 @@
 					function = "spi";
 					groups = "spi0", "spi0_wp_hold";
 				};
+				conf-pu {
+					pins = "SPI0_CSB", "SPI0_HOLD", "SPI0_WP";
+					drive-strength = <MTK_DRIVE_4mA>;
+				};
+				conf-pd {
+					pins = "SPI0_CLK", "SPI0_MOSI", "SPI0_MISO";
+					drive-strength = <MTK_DRIVE_4mA>;
+				};
 			};
 
 			spi1_pins: spi1-pins {
