--- a/target/linux/mediatek/image/filogic.mk
+++ b/target/linux/mediatek/image/filogic.mk
@@ -1200,10 +1200,12 @@ define Device/mediatek_mt7988a-rfb
 	mt7988a-rfb-spim-nand-nmbm \
 	mt7988a-rfb-spim-nor \
 	mt7988a-rfb-eth1-aqr \
+	mt7988a-rfb-eth1-cux3410 \
 	mt7988a-rfb-eth1-i2p5g-phy \
 	mt7988a-rfb-eth1-mxl \
 	mt7988a-rfb-eth1-sfp \
 	mt7988a-rfb-eth2-aqr \
+	mt7988a-rfb-eth2-cux3410 \
 	mt7988a-rfb-eth2-mxl \
 	mt7988a-rfb-eth2-sfp \
 	mt7988a-rfb-spidev
