diff --git a/target/linux/mediatek/filogic/base-files/etc/board.d/02_network b/target/linux/mediatek/filogic/base-files/etc/board.d/02_network
index e015824..2f07aab 100644
--- a/target/linux/mediatek/filogic/base-files/etc/board.d/02_network
+++ b/target/linux/mediatek/filogic/base-files/etc/board.d/02_network
@@ -112,6 +112,10 @@ mediatek_setup_interfaces()
 		ucidef_add_switch "switch0" \
 			"0:lan" "1:lan" "2:lan" "3:lan" "6u@eth0"
 		;;
+	mediatek,mt7987a-spim-nand-cob*)
+		ucidef_set_interfaces_lan_wan "eth0 eth2" eth1
+		;;
+	mediatek,mt7987*|\
 	mediatek,mt7988a-rfb|\
 	mediatek,mt7988d-rfb)
 		ucidef_set_interfaces_lan_wan "lan0 lan1 lan2 lan3 eth2" eth1
