From bc76aa1d770aad3db7b64394606c63e953e11f4b Mon Sep 17 00:00:00 2001
From: Evelyn Tsai <evelyn.tsai@mediatek.com>
Date: Fri, 14 Apr 2023 06:55:22 +0800
Subject: [PATCH 17/17] mac80211: mtk: update max_bssid_indicator based on real
 BSS numbers

---
 net/mac80211/cfg.c | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/net/mac80211/cfg.c b/net/mac80211/cfg.c
index 7449777..b801df7 100644
--- a/net/mac80211/cfg.c
+++ b/net/mac80211/cfg.c
@@ -1073,8 +1073,7 @@ static int ieee80211_assign_beacon(struct ieee80211_sub_if_data *sdata,
 		pos += struct_size(new->mbssid_ies, elem, mbssid->cnt);
 		ieee80211_copy_mbssid_beacon(pos, new->mbssid_ies, mbssid);
 		/* update bssid_indicator */
-		sdata->vif.bss_conf.bssid_indicator =
-			ilog2(__roundup_pow_of_two(mbssid->cnt + 1));
+		sdata->vif.bss_conf.bssid_indicator = *(pos + 2);
 	}
 
 	if (csa) {
-- 
2.18.0

