From 0f379721cbcd19378d22c6ddfa58777da463daf6 Mon Sep 17 00:00:00 2001
From: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
Date: Thu, 16 Nov 2023 13:20:40 +0800
Subject: [PATCH] mac80211: mtk: send deauth frame if CAC is required during
 CSA

Signed-off-by: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
---
 net/mac80211/cfg.c | 30 ++++++++++++++++++++++++++++++
 1 file changed, 30 insertions(+)

diff --git a/net/mac80211/cfg.c b/net/mac80211/cfg.c
index 8a0833b..bee8f69 100644
--- a/net/mac80211/cfg.c
+++ b/net/mac80211/cfg.c
@@ -3355,11 +3355,41 @@ static int __ieee80211_csa_finalize(struct ieee80211_sub_if_data *sdata)
 	struct ieee80211_local *local = sdata->local;
 	u32 changed = 0;
 	int err;
+	bool send_deauth;
 
 	sdata_assert_lock(sdata);
 	lockdep_assert_held(&local->mtx);
 	lockdep_assert_held(&local->chanctx_mtx);
 
+	send_deauth = !cfg80211_chandef_identical(&sdata->vif.bss_conf.chandef,
+						  &sdata->csa_chandef) &&
+		      !cfg80211_reg_can_beacon_relax(local->hw.wiphy,
+						     &sdata->csa_chandef,
+						     sdata->wdev.iftype);
+	/* broadcast deauth frame if CAC is required */
+	if (send_deauth) {
+		struct sta_info *sta, *tmp;
+		u8 frame_buf[IEEE80211_DEAUTH_FRAME_LEN];
+		u8 broadcast[ETH_ALEN] = {0xff, 0xff, 0xff, 0xff, 0xff, 0xff};
+
+		ieee80211_wake_vif_queues(local, sdata, IEEE80211_QUEUE_STOP_REASON_CSA);
+		ieee80211_send_deauth_disassoc(sdata, broadcast,
+					       sdata->vif.bss_conf.bssid,
+					       IEEE80211_STYPE_DEAUTH,
+					       WLAN_REASON_DEAUTH_LEAVING,
+					       send_deauth, frame_buf);
+		ieee80211_stop_vif_queues(local, sdata, IEEE80211_QUEUE_STOP_REASON_CSA);
+
+		mutex_lock(&local->sta_mtx);
+		list_for_each_entry_safe(sta, tmp, &local->sta_list, list) {
+			if (sdata != sta->sdata)
+				continue;
+
+			WARN_ON(__sta_info_destroy(sta));
+		}
+		mutex_unlock(&local->sta_mtx);
+	}
+
 	/*
 	 * using reservation isn't immediate as it may be deferred until later
 	 * with multi-vif. once reservation is complete it will re-schedule the
-- 
2.18.0

