From bb7768975be54f9fb04f04c7851485e80d30dd3a Mon Sep 17 00:00:00 2001
From: Rex Lu <rex.lu@mediatek.com>
Date: Mon, 20 Jan 2025 09:50:05 +0800
Subject: [PATCH] add mtk wed hw version support

---
 include/linux/soc/mediatek/mtk_wed.h | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/include/linux/soc/mediatek/mtk_wed.h b/include/linux/soc/mediatek/mtk_wed.h
index 5ab3a93..14c3914 100644
--- a/include/linux/soc/mediatek/mtk_wed.h
+++ b/include/linux/soc/mediatek/mtk_wed.h
@@ -56,6 +56,14 @@ enum mtk_wed_bus_tye {
 	MTK_WED_BUS_AXI,
 };
 
+enum mtk_wed_hw_version {
+	MTK_WED_DISABLE,
+	MTK_WED_HW_V1,
+	MTK_WED_HW_V2,
+	MTK_WED_HW_V3,
+	MTK_WED_HW_V3_1 = 5,
+};
+
 #define MTK_WED_RING_CONFIGURED		BIT(0)
 struct mtk_wed_ring {
 	struct mtk_wdma_desc *desc;
-- 
2.45.2

