diff --git a/package/network/config/netifd/files/usr/libexec/network/packet-steering.uc b/package/network/config/netifd/files/usr/libexec/network/packet-steering.uc
index 34df46d18ef..5921edbb1f6 100755
--- a/package/network/config/netifd/files/usr/libexec/network/packet-steering.uc
+++ b/package/network/config/netifd/files/usr/libexec/network/packet-steering.uc
@@ -88,7 +88,7 @@ function set_netdev_cpu(dev, cpu, rx_queue) {
 
 function task_device_match(name, device)
 {
-	let napi_match = match(name, /napi\/([^-]*)-\d+/);
+	let napi_match = match(name, /napi\/([^-+])-\d+/);
 	if (!napi_match)
 		napi_match = match(name, /mt76-tx (phy\d+)/);
 	if (napi_match &&

