diff --git a/feeds/packages/net/lftp/Makefile b/feeds/packages/net/lftp/Makefile
index e0ed047..f13296a 100644
--- a/feeds/packages/net/lftp/Makefile
+++ b/feeds/packages/net/lftp/Makefile
@@ -12,8 +12,7 @@ PKG_VERSION:=4.9.2
 PKG_RELEASE:=1
 
 PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.xz
-PKG_SOURCE_URL:=https://lftp.tech/ftp/ \
-		https://mirror.csclub.uwaterloo.ca/gentoo-distfiles/distfiles/
+PKG_SOURCE_URL:= https://sources.cdn.openwrt.org/
 PKG_HASH:=c517c4f4f9c39bd415d7313088a2b1e313b2d386867fe40b7692b83a20f0670d
 
 PKG_MAINTAINER:=Federico Di Marco <fededim@gmail.com>
